title: 'Data Science project: XXX '\
author: "Yaxin Chen, Alewtina Towara, Clara Certa, Linda Kaupp "\
date: "Data Analysis Project SS22"\
supervisors: Dr. Carl Herrmann, Dr. Maria Dinkelacker, Ian Fichter' output: pdf_document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.align = 'centre', message=FALSE)
knitr::opts_knit$set(root.dir = "")
getwd()
```

###load library

```{r load packages, eval=TRUE, include=FALSE, comment=FALSE}
library(affy)
library(vsn)
library(AnnotationDbi)
library(hgu133plus2hsenstcdf)
library(hgu133plus2hsenstprobe)
library(limma)
library(pheatmap)
library(dplyr)
library(tidyverse)
library(hexbin)
library(cluster)
library(factoextra)
library(EnhancedVolcano)
library(org.Hs.eg.db)
library(RColorBrewer)
library(VennDiagram)
```

\newpage

# 1.Abstract

\newpage

# 2.Introduction

\newpage

# 3.Methods and Results

## 3.1 Quality Control and Normalization of the Data

### setting working directory and loading

```{r set working directory}
setwd("~//documents//GitHub//2022-topic-04-team-03//Data//rawData")

data.human=ReadAffy()
data.human@cdfName <- "HGU133Plus2_Hs_ENST"
```

### Single chip control: reading in the microarray chips

```{r chips pictures}
for (i in 1:18){
   image(data.human[,i], col = rainbow (100, start = 0, end = 0.75)[100:1])}
```

### Normalization

```{r normalization, echo=FALSE, message=FALSE, comment=FALSE, results="hide", highlight=TRUE}
human.vsnrma <- vsnrma(data.human)
setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")

expression.data <- exprs(human.vsnrma)
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
#save.image(file="normalized_human_data_18290.rda")#

```

### Plots meanSdPlot

```{r meanSdPlot}
meanSdPlot(human.vsnrma)

setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")
dev.copy2eps(file="meanSdPlot_human_vsnrma_normalized.eps")
```

### Boxplots

```{r boxplots}
names = c("1 cell stage, rep 1", "1 cell stage, rep 2", "1 cell stage, rep 3",
          "2 cell stage, rep 1", "2 cell stage, rep 2", "2 cell stage, rep 3",
          "4 cell stage, rep 1", "4 cell stage, rep 2", "4 cell stage, rep 3",
          "8 cell stage, rep 1", "8 cell stage, rep 2", "8 cell stage, rep 3",
          "morula stage, rep 1", "morula stage, rep 2", "morula stage, rep 3",
          "blastocyst stage, rep 1", "blastocyst stage, rep 2", "blastocyst stage, rep 3")


# Before normalization:
par(las=2)
mmi=c(1,0.7,1.0477939,0.5366749)
par(mai=mmi)
boxplot(data.human, col= rainbow(35), cex.axis=0.5, main="Gene expression in human embroyogenesis data before normalization", xlim = names)

dev.copy2eps(file="boxplot_rawdata_18290.eps")

# After Normalization:
boxplot(exprs(human.vsnrma), col= rainbow(35), cex.axis=0.5, main="Gene expression in human embroyogenesis data after normalization")

dev.copy2eps(file="boxplot_normalizeddata_18290.eps")
dev.off() # what does this function do?
```

### Density Plots

```{r density plots}
# Before Normalization:
hist(data.human, col=rainbow(35), main="Density function of log Intensity of human embryogenesis raw data")

dev.copy2pdf(file="Histogram_rawdata_39897.pdf", width = 10, height = 7)

# After Normalization:
eset=exprs(human.vsnrma)

plot(density(eset[,1]), type="n", xlab="log Intensity", ylim=c(0,1), main="Density function of log Intensity of human embryogenesis normalized")
for (i in 1:ncol(eset)) {
  lines(density(eset[,i]), col=rainbow(35)[i])
}

dev.copy2pdf(file="Histogram_NormalizedData_39897.pdf", width = 10, height = 7)
```

### RNA Degradation Plot

```{r RNA Degradation Plot}
par(mfrow = c(2,1))
rnadeg.raw = AffyRNAdeg(data.human)

plotAffyRNAdeg(rnadeg.raw, col=rainbow(35))
title(sub="human embroyogenesis rawdata")

plotAffyRNAdeg(rnadeg.raw, col=rainbow(35), transform= "shift.only")
title(sub="human embryogenesis rawdata - shifted")

dev.copy2pdf(file="RNAdegrad_plot.pdf", width = 12.5, height = 20)
#dev.off()
```

### Scatter Plot

```{r scatter plot}
setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")
expression.data <- exprs(human.vsnrma)

for(i in 1:17){
  plot(expression.data[,c(i,i+1)], pch=".", cex=2)
  abline(0, 1, col="red")               # 45 degree dividing line
  
  title(main = paste("Scatterplot of probe", 
                     substr(colnames(human.vsnrma)[i], 1, nchar(colnames(human.vsnrma)[i])), "and", 
                     substr(colnames(human.vsnrma)[i+1], 1, nchar(colnames(human.vsnrma)[i+1])), 
                     sep=" ", collapse = NULL))
  
  file.name <- paste("~//documents//GitHub//2022-topic-04-team-03//Plots", 
                     as.character(substr(colnames(human.vsnrma)[i], 1, nchar(colnames(human.vsnrma)[i]))), "_",
                     as.character(substr(colnames(human.vsnrma)[i+1], 1, nchar(colnames(human.vsnrma)[i+1]))),
                     ".pdf", sep="")
  dev.copy2pdf(file = file.name)
  dev.off()
}
```

### Data Clean-up

```{r data clean-up}
# Are there any NAs in our dataset?
sum(apply(human.vsnrma.df,1,function(x){sum(is.na(x))}))
#> No

```

## 3.2 Data Analysis

### Annotation

```{r }
# read in ensembl table with following attributes: "Gene stable ID", 
#"Gene stable ID version", "Transcript stable ID", "Transscript stable ID version", 
#"Gene.name", "Transcript name", "Chromosome.scaffold.name", "Gene.description", "HGNC.symbol"

setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")

ensembl.data <- read.csv("ensembl.human.txt")
dim(ensembl.data)
##268341      9
```

```{r}
# create variables that contain gene ID, transcript ID, 
#the chromosome name and the gene symbol
ensembl.genes = ensembl.data[,1]
ensembl.transcripts = ensembl.data[,3]
ensembl.chromosome = ensembl.data[,7]
ensembl.symbols = ensembl.data[,9]
```

```{r}
# Create a data frame out of the expression data from the normalized data
human.vsnrma.df = data.frame(exprs(human.vsnrma))
```

```{r}
#rename the colnames to the stages
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456643.CEL'] <- '1-cell stage.rep.1'
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456644.CEL'] <- '1-cell stage.rep.2' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456645.CEL'] <- '1-cell stage.rep.3' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456646.CEL'] <- '2-cell stage.rep.1' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456647.CEL'] <- '2-cell stage.rep.2' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456648.CEL'] <- '2-cell stage.rep.3' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456649.CEL'] <- '4-cell stage.rep.1' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456650.CEL'] <- '4-cell stage.rep.2' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456651.CEL'] <- '4-cell stage.rep.3' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456652.CEL'] <- '8-cell stage.rep.1' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456653.CEL'] <- '8-cell stage.rep.2' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456654.CEL'] <- '8-cell stage.rep.3' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456655.CEL'] <- 'morula stage.rep.1' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456656.CEL'] <- 'morula stage.rep.2' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456657.CEL'] <- 'morula stage.rep.3'
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456658.CEL'] <- 'blastocyst stage.rep.1' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456659.CEL'] <- 'blastocyst stage.rep.2' 
names(human.vsnrma.df)[names(human.vsnrma.df) == 'GSM456660.CEL'] <- 'blastocyst stage.rep.3' 
```

```{r}
# Check dimensions
dim(human.vsnrma.df)
#95,721    18
# exclude Affymetrix control genes which begin with "AFFX"
human.vsnrma.df2 = human.vsnrma.df[63:95721,]

dim(human.vsnrma.df2)
# 95,659    18
```

```{r}
# remove ".xx_at" from the rownames
rownames(human.vsnrma.df2) = substr(rownames(human.vsnrma.df2), 1,15)

# read in TRA data for human
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
tra.data <- read.table("tra.2017.human.gtex.5x.table.tsv",header=TRUE)
dim(tra.data)
#60131    10
```

```{r}
# find the index of rownames of our dataset that contain transcript Ids from the tra dataset
# We extracted the TRA genes out of the three tables(human.vsnrma, tra, ensembl)

j = which(rownames(human.vsnrma.df2) %in% tra.data$ensembl.transcript) 
tra.extracted = rownames(human.vsnrma.df2)[j] #24,783
human.vsnrma.only.tra = human.vsnrma.df2[j,]
#dim(24783,18)

k = which(tra.data$ensembl.transcript %in% tra.extracted)
tra.expressed.in.chips = tra.data[k,] #24,783 
#dim(24783,10)
```

150 genes weniger als bei anderen zwei Tabellen

```{r}
c = which(ensembl.transcripts %in% tra.extracted)
ensembl.only.tra = ensembl.data[c,]
#dim(24623,9)
```

```{r}
#reorder the rows of ensembl.only.tra and tra.expressed.in.chips
tra.expressed.in.chips = arrange(tra.expressed.in.chips,ensembl.transcript)
ensembl.only.tra = arrange(ensembl.only.tra,Transcript.stable.ID)

#fusion of two tables! tra.expressed.in.chips and human.vsnrma.only.tra
fusion.tra.expression.tra.table = cbind(human.vsnrma.only.tra,tra.expressed.in.chips$ensembl.gene,tra.expressed.in.chips$tiss.number,tra.expressed.in.chips$tissues,tra.expressed.in.chips$max.tissue)
colnames(fusion.tra.expression.tra.table)[19:22]<-c("ensembl.gene","tissue.number","tissues","max.tissue")


# fusion of fusion.tra.expression.tra.table and ensembl.only.tra because ensembl.new contains less rows than the other two tables
p = which(rownames(fusion.tra.expression.tra.table) %in% ensembl.transcripts)
fusion.tra.expression.tra.table.extracted = fusion.tra.expression.tra.table[p,]
fusion.tra.expression.tra.table.ensembl.table = cbind(fusion.tra.expression.tra.table.extracted,ensembl.only.tra$Gene.name,ensembl.only.tra$Gene.description,ensembl.only.tra$Chromosome.scaffold.name)

#rename colnames
colnames(fusion.tra.expression.tra.table.ensembl.table)[23:25]<-c("Gene.name","Gene.description","Chromosome.scaffold.name")


setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
write.csv(fusion.tra.expression.tra.table.ensembl.table, file="fusion.tra.expression.tra.table.ensembl.table.csv")

```

### Exploratory data analysis

```{r}
# Generate a list containing all tissue names with number of tissue TRAs, which is detected in our chips
tissue = c("Brain","Esophagus","Heart","Liver","Cervix","Muscle","Ovary","Colon","Breast","Kidney","Pituitary","Testis","Whole Blood","Cells - Transformed fibroblasts", "Cells - EBV-transformed lymphocytes", "Spleen","Lung","Stomach","Pancreas","Small Intestine","Skin","Artery","Adrenal","Gland","Pituitary","Nerve","Minor","Salivary","Bladder","Adipose","Thyroid","Prostate","Vagina")
tissue.number = c()

for (i in 1:33) {
  tissue.number[[i]] = nrow(fusion.tra.expression.tra.table.ensembl.table %>% filter(grepl(tissue[i],fusion.tra.expression.tra.table.ensembl.table[,21])))
}
print(tissue.number)
tissue.distribution = cbind(tissue,tissue.number)
tissue.distribution = as.data.frame(tissue.distribution)
```

```{r}
# Arranging the list according to the descending TRAs number
tissue.distribution.arranged = arrange(tissue.distribution, desc(as.numeric(tissue.number)))
```

```{r}
# Create a barplot
coul <- brewer.pal(5, "Set2")
barplot(height = as.numeric(tissue.distribution.arranged$tissue.number[1:10]),names.arg = tissue.distribution.arranged$tissue[1:10],cex.names=0.8,col=coul, main="Frequency of TRAs in our data", ylab="Frequency")

setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")

dev.copy2pdf(file="tissue.distribution.arranged.barplot.pdf" )
```

```{r}
# Create a heatmap of the TRA genes in our dataset
## What does unlist mean? -> Transforms list into matrix
human.vsnrma.only.tra.matrix = matrix(unlist(human.vsnrma.only.tra), ncol = 6, nrow = 24783)

colnames(human.vsnrma.only.tra.matrix) = c("1 cell stage", "2 cell stage", "4 cell stage", "8 cell stage", "morula stage", "blastocyst stage")

heatmap = pheatmap(human.vsnrma.only.tra.matrix, 
                   cluster_cols=FALSE, 
                   show_rownames = TRUE, 
                   legend=TRUE, 
                   fontsize_row=0.5,
                   main = "Expression of 24,783 TRA genes in 6 different stages of embryogenesis")

setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")
dev.copy2pdf(file="heatmap.pdf")

```

### Principal component analysis

```{r code von Carl}
###Using PCA to analyse our 18 chips, to see whether there are any outlier chip and whether the different replicates cluster with each other.

topVar = apply(human.vsnrma.df2, 1, var)
q75 = quantile(topVar, probs = 0.75)
i.topvar = which(topVar >= q75)
human.vsnrma.df2.topVar = human.vsnrma.df2[i.topvar,]
dim(human.vsnrma.df2.topVar)

pca = prcomp(t(human.vsnrma.df2.topVar), center = T, scale. = T)
print(pca)
```

```{r show how much variance each PC explains}
plot(pca$sdev)
variance = (pca$sdev)^2
prop.variance = variance/sum(variance)
```

```{r}
# Plot cumulative proportion of variance explained
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
```

4 PCs are enough to show more than 90%

```{r Plot PC1 and PC2}
color = c(rep("red",3),rep("orange",3),rep("yellow",3),rep("green",3),rep("blue",3),rep("purple",3))

plot(pca$x[,1], pca$x[,2],col=color, pch=19,xlab='PC1',ylab='PC2')
```

We cannot observe any outlier replicate and we can see four cluster (Cluster 1:1-cell,2-cell and 4-cell stage, Cluster 2: 8-cell stage, Cluster 3:Morula, Cluster 4:Blastocyst)

```{r plots PC2/3 and PC3/4}
#Plot PC2 and PC3
plot(pca$x[,2], pca$x[,3],col=color, pch=19,xlab='PC2',ylab='PC3')

#Plot PC3 and PC4
plot(pca$x[,3], pca$x[,4],col=color, pch=19,xlab='PC3',ylab='PC4')
```

Now we want to use Kmeans on the first 4 PCs (which explain 90% of the Variance) to find potential clusters and compare it with the cluster that we have observed in PCA just now

```{r run models with varying value of k}
#Use map_dbl to run many models with varying value of k (centers)
tot_withinss <- map_dbl(1:10,  function(k){
  model <- kmeans(x = pca$x[,1:4], centers = k,nstart=30)
  model$tot.withinss
})
```

```{r data frame with k & tot_withinns}
# Generate a data frame containing both k and tot_withinss
elbow_df <- data.frame(
  k = 1:10 ,
  tot_withinss = tot_withinss
)
```

```{r plot with elbow plot}
ggplot(elbow_df, aes(x = k, y =tot_withinss)) +
  geom_line() +
  scale_x_continuous(breaks = 1:10)
```

seems like k=2 is the best

```{r show cluster}
model <- kmeans(x = pca$x[,1:4], centers = 2,nstart=30)
model$cluster
```

Elbow plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)

Plot the Siloutte plo

```{r Use map_dbl to run many models with varying value of k}
sil_width <- map_dbl(2:10,  function(k){
  model <- pam(pca$x[,1:4], k = k)
  model$silinfo$avg.width
})
```

```{r}
# Generate a data frame containing both k and sil_width
sil_df <- data.frame(
  k = 2:10,
  sil_width = sil_width
)
```

```{r}
# Plot the relationship between k and sil_width
ggplot(sil_df, aes(x = k, y = sil_width)) +
  geom_line() +
  scale_x_continuous(breaks = 2:10)
```

it seems like that k=4 is a good choice

```{r}
# Generate a k-means model using the pam() function with a k = 4
pam_k4 <- pam(pca$x[,1:13],k=4)
```

```{r}
# Show the clustering
pam_k4$clustering
```

Silouette Plot suggests 4 Clusters: Cluster 1:1-cell,2-cell and 4-cell stage, Cluster 2: 8-cell stage, Cluster 3:Morula, Cluster 4:Blastocyst)

Using Kmeans based on the original expression data (human.vsnrma.df2) to cluster and compare the results with the Kmeans results based on 4PCs

```{r}
# Use map_dbl to run many models with varying value of k (centers)
tot_withinss <- map_dbl(1:10,  function(k){
    model <- kmeans(x = t(human.vsnrma.df2), centers = k,nstart=30)
    model$tot.withinss
})
```

```{r}
# Generate a data frame containing both k and tot_withinss
elbow_df <- data.frame(
    k = 1:10 ,
    tot_withinss = tot_withinss
)
```

```{r}
# Plot the elbow plot
ggplot(elbow_df, aes(x = k, y =tot_withinss)) +
    geom_line() +
    scale_x_continuous(breaks = 1:10)
```

seems like k=2 is the best

```{r}
#show cluster
model <- kmeans(x = t(human.vsnrma.df2), centers = 2,nstart=30)
model$cluster
```

Elbow plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)


Plot the Siloutte plot

```{r}
# Use map_dbl to run many models with varying value of k
sil_width <- map_dbl(2:10,  function(k){
    model <- pam(t(human.vsnrma.df2), k = k)
    model$silinfo$avg.width
})
```

```{r}
# Generate a data frame containing both k and sil_width
sil_df <- data.frame(
    k = 2:10,
    sil_width = sil_width
)
```

```{r}
# Plot the relationship between k and sil_width
ggplot(sil_df, aes(x = k, y = sil_width)) +
    geom_line() +
    scale_x_continuous(breaks = 2:10)
```

it seems like that k=2 is a good choice

```{r}
# Generate a k-means model using the pam() function with a k = 4
pam_k2 <- pam(t(human.vsnrma.df2),k=2)
  
# Show the clustering
pam_k2$clustering
```

Silouette plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)

Finding genes which explain for the variance by looking at genes with the highst loading in PCs

Use PCA to plot 3 replicates of 1-cell stage and 8-cell stage

```{r}
a=c(1,2,3,10,11,12)
pca.1.8=prcomp(t(human.vsnrma.df2[,a]))
```

```{r}
# Plot cumulative proportion of variance explained
variance = (pca.1.8$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
```

two PCs are enough to show 90% variance

Finding variance genes which explain the variance of PC1 between 1-cell stage and 8-cell stage

```{r}
pc1=pca.1.8$rotation[,1]
top_genes_pca1 = names(sort(abs(pc1), decreasing = T)[1:10])

#Variance genes 1-cell stage and 8-cell stage
j = which(ensembl.data$Transcript.stable.ID%in% top_genes_pca1)
View(ensembl.data[j,])

#Variance TRA genes 1-cell stage and 8-cell stage
k = which(rownames(fusion.tra.expression.tra.table.ensembl.table)%in% top_genes_pca1)
fusion.tra.expression.tra.table.ensembl.table[k,]
```

```{r}
#Use PCA to plot 3 replicates of 8-cell stage and morula stage
pca.8.m=prcomp(t(human.vsnrma.df2[,10:15]))
```

```{r}
# Plot cumulative proportion of variance explained
plot(pca.8.m$sdev)
variance = (pca.8.m$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
```

three PCs are enough to show 90% variance

Finding variance genes which explain the variance of PC1 between 8-cell stage and morula stage

```{r}
pc1=pca.8.m$rotation[,1]
top_genes_pca1 = names(sort(abs(pc1), decreasing = T)[1:10])

#Variance genes 8-cell stage and morula stage
j = which(ensembl.data$Transcript.stable.ID%in% top_genes_pca1)
View(ensembl.data[j,])

#Variance TRA genes 8-cell stage and morula stage
k = which(rownames(fusion.tra.expression.tra.table.ensembl.table)%in% top_genes_pca1)
fusion.tra.expression.tra.table.ensembl.table[k,]
```

*Use PCA to plot 3 replicates of 8-cell stage and blastocyst stage*

Use PCA to plot 3 replicates of morula stage and blastocyst stage

```{r}
pca.m.b=prcomp(t(human.vsnrma.df2[,13:18]))

# Plot cumulative proportion of variance explained
plot(pca.m.b$sdev)
variance = (pca.m.b$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
```

four PCs are enough to show 90% variance

Finding variance genes which explain the variance of PC1 between morula stage and blastocyst stage

```{r}
pc1=pca.m.b$rotation[,1]
top_genes_pca1 = names(sort(abs(pc1), decreasing = T)[1:10])

#Variance genes morula stage and blastocyst stage
j = which(ensembl.data$Transcript.stable.ID%in% top_genes_pca1)
ensembl.data[j,]

#Variance tra genes morula stage and blastocyst stage
k = which(rownames(fusion.tra.expression.tra.table.ensembl.table)%in% top_genes_pca1)
fusion.tra.expression.tra.table.ensembl.table[k,]
```

### Limma analysis

```{r limma Analysis}
names = c("1 cell stage, rep 1", "1 cell stage, rep 2", "1 cell stage, rep 3",
          "2 cell stage, rep 1", "2 cell stage, rep 2", "2 cell stage, rep 3",
          "4 cell stage, rep 1", "4 cell stage, rep 2", "4 cell stage, rep 3",
          "8 cell stage, rep 1", "8 cell stage, rep 2", "8 cell stage, rep 3",
          "morula stage, rep 1", "morula stage, rep 2", "morula stage, rep 3",
          "blastocyst stage, rep 1", "blastocyst stage, rep 2", "blastocyst stage, rep 3")


# Rename colnames of our dataset: "x stage, rep y" for a better overview
colnames(human.vsnrma.df2) = names

# Define the different stages of the chips in a vector
stage = factor(x= c(1,1,1,2,2,2,4,4,4,8,8,8,"morula","morula","morula", "blastocyst", "blastocyst", "blastocyst"),levels= c(1,2,4,8,"morula","blastocyst"))

# Create a design matrix with the stages
design = model.matrix(~0 + stage)
            
# Create a contrast matrix which compares all the stages with one another 
cm = makeContrasts(stage1.2 = stage1-stage2, stage1.4 = stage1-stage4, stage1.8 = stage1-stage8, stage1.morula = stage1-stagemorula, stage1.blastocyst = stage1-stageblastocyst,
                   stage2.4 = stage2-stage4, stage2.8 = stage2-stage8, stage2.morula = stage2-stagemorula,  stage2.blastocyst = stage2-stageblastocyst,
                   stage4.8 = stage4-stage8, stage4.morula = stage4-stagemorula,  stage4.blastocyst = stage4-stageblastocyst,
                   stage8.morula = stage8-stagemorula, stage8.blastocyst = stage8-stageblastocyst,
                   stagemorula.blastocyst = stagemorula-stageblastocyst, 
                   levels = design)

# Fit the coefficients of the model
fit = lmFit(human.vsnrma.df2, design)
fit2 = contrasts.fit(fit, contrasts = cm)

# Calculate the t-statistics
fit2 = eBayes(fit2)

# Number of differentially expressed genes (p-value = 0.05)
results = decideTests(fit2, p.value = 0.05)
summary(results)

# Number of differentially expressed genes (p-value = 0.01)
results2 = decideTests(fit2, p.value = 0.01)
summary(results2)

# Create fits for every contrast and then create tables with the results of Limma analysis
# Stage 1-2
fit.1.2 = contrasts.fit(fit, contrasts = cm[,1])
fit.1.2 = eBayes(fit.1.2)

pvalue01 = sum(p.adjust(fit.1.2$p.value, "BH") < 0.01)
# 0
pvalue05 = sum(p.adjust(fit.1.2$p.value, "BH") < 0.05)
# 0

```

```{r fitFunction}
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")

fit.function = function(i){
  fit.table = contrasts.fit(fit, contrasts = cm[,i])
  fit.table = eBayes(fit.table)
}
```

```{r limma fit tables}
fit.1.2 = fit.function(1)
fit.1.4 = fit.function(2)
fit.1.8 = fit.function(3)
fit.1.m = fit.function(4)
fit.1.b = fit.function(5)
fit.2.4 = fit.function(6)
fit.2.8 = fit.function(7)
fit.2.m = fit.function(8)
fit.2.b = fit.function(9)
fit.4.8 = fit.function(10)
fit.4.m = fit.function(11)
fit.4.b = fit.function(12)
fit.8.m = fit.function(13)
fit.8.b = fit.function(14)
fit.m.b = fit.function(15)
```

```{r topTable function}
top.table = function(i){
  fit.table = contrasts.fit(fit, contrasts = cm[,i])
  fit.table = eBayes(fit.table)
  
  pvalue01 = sum(p.adjust(fit.table$p.value, "BH") < 0.01)
  topTable(fit.table, number = pvalue01)
}
```

```{r limma topTables}
limma.table.1.2 = top.table(1)
limma.table.1.4 = top.table(2)
limma.table.1.8 = top.table(3)
limma.table.1.m = top.table(4)
limma.table.1.b = top.table(5)
limma.table.2.4 = top.table(6)
limma.table.2.8 = top.table(7)
limma.table.2.m = top.table(8)
limma.table.2.b = top.table(9)
limma.table.4.8 = top.table(10)
limma.table.4.m = top.table(11)
limma.table.4.b = top.table(12)
limma.table.8.m = top.table(13)
limma.table.8.b = top.table(14)
limma.table.m.b = top.table(15)
```

```{r annotation function}
limma.annotation = function(x){
a = which(ensembl.transcripts %in% rownames(x))
ensembl.new = ensembl.data[a,]
  
b = which(rownames(x) %in% ensembl.new$Transcript.stable.ID)
x.new = x[b,]
  
ensembl.new = arrange(ensembl.new,Transcript.stable.ID)
x.new = arrange(x.new, rownames(x.new))
  
annotated.x = cbind(x.new, ensembl.new$Gene.name, ensembl.new$Gene.description,ensembl.new$Chromosome.scaffold.name, ensembl.new$Gene.stable.ID)
colnames(annotated.x)[7:10]<-c("gene.name","gene.description","Chromosome.scaffold.name", "Gene.stable.ID")
annotated.x = arrange(annotated.x, adj.P.Val)
  return(annotated.x)
}
```

```{r limma annotation}
annotated.limma.1.8 = limma.annotation(limma.table.1.8)
annotated.limma.1.m = limma.annotation(limma.table.1.m)
annotated.limma.1.b = limma.annotation(limma.table.1.b)
#annotated.limma.2.4 = limma.annotation(limma.table.2.4)
annotated.limma.2.8 = limma.annotation(limma.table.2.8)
annotated.limma.2.m = limma.annotation(limma.table.2.m)
annotated.limma.2.b = limma.annotation(limma.table.2.b)
annotated.limma.4.8 = limma.annotation(limma.table.4.8)
annotated.limma.4.m = limma.annotation(limma.table.4.m)
annotated.limma.4.b = limma.annotation(limma.table.4.b)
annotated.limma.8.m = limma.annotation(limma.table.8.m)
annotated.limma.8.b = limma.annotation(limma.table.8.b)
annotated.limma.m.b = limma.annotation(limma.table.m.b)
```

```{r limma TRAs}
#annotated limma tables mit TRAs annotaten
#1.4 und 1.2 sind 00
annotated.limma.1.8$ensembl.transcript = rownames(annotated.limma.1.8)
annotated.limma.1.8.tra <- merge(annotated.limma.1.8, tra.data, by = 'ensembl.transcript')

annotated.limma.1.m$ensembl.transcript = rownames(annotated.limma.1.m)
annotated.limma.1.m.tra <- merge(annotated.limma.1.m, tra.data, by = 'ensembl.transcript')

annotated.limma.1.b$ensembl.transcript = rownames(annotated.limma.1.b)
annotated.limma.1.b.tra <- merge(annotated.limma.1.b, tra.data, by = 'ensembl.transcript')
#2.4. ist 00
annotated.limma.2.8$ensembl.transcript = rownames(annotated.limma.2.8)
annotated.limma.2.8.tra <- merge(annotated.limma.2.8, tra.data, by = 'ensembl.transcript')

annotated.limma.2.m$ensembl.transcript = rownames(annotated.limma.2.m)
annotated.limma.2.m.tra <- merge(annotated.limma.2.m, tra.data, by = 'ensembl.transcript')

annotated.limma.2.b$ensembl.transcript = rownames(annotated.limma.2.b)
annotated.limma.2.b.tra <- merge(annotated.limma.2.b, tra.data, by = 'ensembl.transcript')

annotated.limma.4.8$ensembl.transcript = rownames(annotated.limma.4.8)
annotated.limma.4.8.tra <- merge(annotated.limma.4.8, tra.data, by = 'ensembl.transcript')

annotated.limma.4.m$ensembl.transcript = rownames(annotated.limma.4.m)
annotated.limma.4.m.tra <- merge(annotated.limma.4.m, tra.data, by = 'ensembl.transcript')

annotated.limma.4.b$ensembl.transcript = rownames(annotated.limma.4.b)
annotated.limma.4.b.tra <- merge(annotated.limma.4.b, tra.data, by = 'ensembl.transcript')

annotated.limma.8.m$ensembl.transcript = rownames(annotated.limma.8.m)
annotated.limma.8.m.tra <- merge(annotated.limma.8.m, tra.data, by = 'ensembl.transcript')

annotated.limma.8.b$ensembl.transcript = rownames(annotated.limma.8.b)
annotated.limma.8.b.tra <- merge(annotated.limma.8.b, tra.data, by = 'ensembl.transcript')

annotated.limma.m.b$ensembl.transcript = rownames(annotated.limma.m.b)
annotated.limma.m.b.tra <- merge(annotated.limma.m.b, tra.data, by = 'ensembl.transcript')

```


```{r tissue plot}
#Erstellen Means pro stage pro tissue
n_occur1.8=data.frame(table(annotated.limma.1.8.tra$max.tissue))
tissue.vector1.8=c()
for (i in 1:53) {

  tissue.genes=which(annotated.limma.1.8.tra$max.tissue %in% n_occur1.8$Var1[i])
  tissue.vector1.8[[i]]= mean(c(annotated.limma.1.8.tra[tissue.genes,2]))
}


n_occur1.m=data.frame(table(annotated.limma.1.m.tra$max.tissue))
tissue.vector1.m=c()
for (i in 1:53) {

  tissue.genes=which(annotated.limma.1.m.tra$max.tissue %in% n_occur1.m$Var1[i])
  tissue.vector1.m[[i]]= mean(c(annotated.limma.1.m.tra[tissue.genes,2]))
}

n_occur1.b=data.frame(table(annotated.limma.1.b.tra$max.tissue))
tissue.vector1.b=c()
for (i in 1:53) {

  tissue.genes=which(annotated.limma.1.b.tra$max.tissue %in% n_occur1.b$Var1[i])
  tissue.vector1.b[[i]]= mean(c(annotated.limma.1.b.tra[tissue.genes,2]))
}
tissue.vector1.2=c()
tissue.vector1.2[1:53] =0
tissue.vector1.4=c()
tissue.vector1.4[1:53] =0
#Erstellen Plot Data
tissue.matrix = cbind(tissue.vector1.2,tissue.vector1.4,tissue.vector1.8,tissue.vector1.m,tissue.vector1.b)
rownames(tissue.matrix) = n_occur1.8$Var1[1:53] 
tissue.matrix= cbind(rownames(tissue.matrix),tissue.matrix)
tissue.matrix= as_tibble(tissue.matrix)

tissue.matrix=pivot_longer(tissue.matrix, cols=tissue.vector1.2:tissue.vector1.b, names_to = "stages" ,values_to="logFc")

#plot# Fragggeee???
ggplot(tissue.matrix, aes(x = stages, y= as.numeric(logFc),group=as.character(V1)))  + geom_line(aes(color=as.character(V1))) +
           geom_point(aes(color=as.character(V1))) + labs(y=("logFC")) + labs(colour = "tissues")

```

### filter DE genes
create Volcanoplots to visualize the DE genes
using Enhancedvolcanoplot with input from limma tables that contain all genes

we found that the ensembl table has some transcripts that repeats itself, so we only select the unique transcripts from the ensembl data

```{r}
n_occur=data.frame(table(ensembl.data$Transcript.stable.ID))
n_occur=arrange(n_occur,desc(n_occur$Freq))
a=sum(n_occur$Freq>1)
duplicate.ensembl=n_occur$Var1[1:40]
j=which(ensembl.data$Transcript.stable.ID%in%duplicate.ensembl)  
ensembl.duplicate=ensembl.data[j,]
ensembl.duplicate=arrange(ensembl.duplicate,ensembl.duplicate$Transcript.stable.ID)
Nth.delete<-function(dataframe, n)dataframe[-(seq(n,to=nrow(dataframe),by=n)),]
ensembl.duplicate.genes=Nth.delete(ensembl.duplicate, 2)
ensembl.unique=rbind(ensembl.data[-j,],ensembl.duplicate.genes)
```


generate a function with the output of limma tables (complete) with sig. genes and non sig. genes
```{r limma tables with sig & non sig. genes}
generate.limma.table.vollst = function(fit,n){
  fit.1 = contrasts.fit(fit, contrasts = cm[,n])
  fit.1 = eBayes(fit.1)
  limma.table.vollst=topTable(fit.1,number = dim(fit.1)[1])
  return(limma.table.vollst)
}
limma.table.vollst.1.2=generate.limma.table.vollst(fit,1)
limma.table.vollst.1.4=generate.limma.table.vollst(fit,2)
limma.table.vollst.1.8=generate.limma.table.vollst(fit,3)
limma.table.vollst.1.m=generate.limma.table.vollst(fit,4)
limma.table.vollst.1.b=generate.limma.table.vollst(fit,5)
limma.table.vollst.2.4=generate.limma.table.vollst(fit,6)
limma.table.vollst.2.8=generate.limma.table.vollst(fit,7)
limma.table.vollst.2.m=generate.limma.table.vollst(fit,8)
limma.table.vollst.2.b=generate.limma.table.vollst(fit,9)
limma.table.vollst.4.8=generate.limma.table.vollst(fit,10)
limma.table.vollst.4.m=generate.limma.table.vollst(fit,11)
limma.table.vollst.4.b=generate.limma.table.vollst(fit,12)
limma.table.vollst.8.m=generate.limma.table.vollst(fit,13)
limma.table.vollst.8.b=generate.limma.table.vollst(fit,14)
limma.table.vollst.m.b=generate.limma.table.vollst(fit,15)
```

anotation of the limma table (complete) with emsembl information
```{r}
ensembl.limma.annotation = function(x){}
#=======

ensembl.limma.annotation <- function(x){
  a = which(ensembl.unique$Transcript.stable.ID %in% rownames(x))
  ensembl.new = ensembl.unique[a,]
  
  b = which(rownames(x)%in% ensembl.new$Transcript.stable.ID)
  x.new = x[b,]
  
  ensembl.new = arrange(ensembl.new,ensembl.new$Transcript.stable.ID)
  x.new = arrange(x.new, rownames(x.new))
  
  annotated.x = cbind(x.new, ensembl.new$Gene.name, ensembl.new$Gene.description,ensembl.new$Chromosome.scaffold.name)
  colnames(annotated.x)[7:9]<-c("gene.name","gene.description","Chromosome.scaffold.name")
  annotated.x = arrange(annotated.x, adj.P.Val)
  return(annotated.x)
}

annotated.limma.table.vollst.1.8 = ensembl.limma.annotation(limma.table.vollst.1.8)
annotated.limma.table.vollst.8.m = ensembl.limma.annotation(limma.table.vollst.8.m) 
annotated.limma.table.vollst.m.b = ensembl.limma.annotation(limma.table.vollst.m.b)
```

```{r volcano plots}
EnhancedVolcano(annotated.limma.table.vollst.1.8[,1:6],lab = annotated.limma.table.vollst.1.8[,7],title = "1-cell stadium vs. 8-cell stadium",x = 'logFC',xlab = 'logFC',y = 'adj.P.Val', ylab = '-log(adj.P.Val)',pCutoff = 0.01,
                FCcutoff = 4, pointSize = 1.5,labSize = 2.5, legendLabels = c('Not sig.','LogFC','adj.P.value','p-value & LogFC'))

EnhancedVolcano(annotated.limma.table.vollst.8.m[,1:6],lab = annotated.limma.table.vollst.8.m[,7],title = "8-cell stadium vs. morula stadium",x = 'logFC',xlab = 'logFC',y = 'adj.P.Val', ylab = '-log(adj.P.Val)', pCutoff = 0.01,
                FCcutoff = 3, pointSize = 1.5,labSize = 2.5, legendLabels = c('Not sig.','LogFC','adj.P.value','p-value & LogFC'))

EnhancedVolcano(annotated.limma.table.vollst.m.b[,1:6],lab = annotated.limma.table.vollst.m.b[,7],title = "morula stadium vs. blastocyst stadium",x = 'logFC',xlab = 'logFC',y = 'adj.P.Val',ylab = '-log(adj.P.Val)', pCutoff = 0.01,
                FCcutoff = 3, pointSize = 1.5,labSize = 2.5, legendLabels = c('Not sig.','LogFC','adj.P.value','p-value & LogFC'))
```

annotate the tra genes in limma tables (only sig. genes) with tra information

```{r}
fuse.tra.limma <- function(x){
  
  x = arrange(x,rownames(x))
  j = which(rownames(x) %in% tra.data$ensembl.transcript) 
  
  tra.extracted.stages = rownames(x)[j] 
  
  tra.stages = x[j,]
  
  k = which(tra.data$ensembl.transcript %in% tra.extracted.stages)
  tra.data.stages = tra.data[k,]
  tra.data.stages = arrange(tra.data.stages,tra.data.stages$ensembl.transcript)
  
  tra.limma.stages = cbind(tra.stages,tra.data.stages$ensembl.gene,tra.data.stages$tiss.number,tra.data.stages$tissues,tra.data.stages$max.tissue)
  colnames(tra.limma.stages)[10:13]<-c("ensembl.gene","tissue.number","tissues","max.tissue")
  
  
  x = cbind(x,"ensembl.gene","tissue.number","tissues","max.tissue")
  x[,10:13] = "NA"
  colnames(x)[10:13] = c("ensembl.gene","tissue.number","tissues","max.tissue")
  annotated.tra.limma.stages = rbind(tra.limma.stages,x[-j,])
  annotated.tra.limma.stages = arrange(annotated.tra.limma.stages,desc(annotated.tra.limma.stages$logFC))
  return(annotated.tra.limma.stages)
}

```

```{r}
###1.8-stadium
annotated.tra.limma.1.8 = fuse.tra.limma(annotated.limma.1.8)
###8.m-stadium  
annotated.tra.limma.8.m = fuse.tra.limma(annotated.limma.8.m) 
###m.b-stadium
annotated.tra.limma.m.b = fuse.tra.limma(annotated.limma.m.b)

View(annotated.tra.limma.1.8)
```

filter the top 20 DE upregulated genes for diffrent stages:
```{r 1 cell & 8 cell}
  #abline(v=mean(abs(limma.table.1.8$logFC)))#
  #c=which(limma.table.1.8$logFC< 0)
  #limma.1.8.down=limma.table.1.8[c,]
  #limma.1.8.down.sig=rownames(annotated.limma.1.8.down[which(annotated.limma.1.8.down$P.Value < 0.05),])
  #limma.table.1.8.arranged=arrange(limma.table.1.8,desc(abs(logFC)))
  #limma.1.8.down.sig=arrange(limma.1.8.down.sig,)
 hist(annotated.limma.1.8$logFC)
c=which(annotated.limma.1.8$logFC > 2)
limma.table.1.8.up=annotated.limma.1.8[c,]
limma.1.8.up.sig=limma.table.1.8.up[which(limma.table.1.8.up$P.Value < 0.05),]
limma.1.8.up.sig=arrange(limma.1.8.up.sig,desc(limma.1.8.up.sig$logFC))  
interesting_genes.1.8=rownames(limma.1.8.up.sig[1:20,])

k = which(rownames(fusion.tra.expression.tra.table.ensembl.table)%in% interesting_genes.1.8)
  View(fusion.tra.expression.tra.table.ensembl.table[k,])
  
j = which(ensembl.data$Transcript.stable.ID%in%interesting_genes.1.8 )
  View(ensembl.data[j,])
```

```{r between 8 cell & m}
hist(limma.table.8.m$logFC)

  
  #c=which(limma.table.8.m$logFC< -3)
  #limma.table.8.m.down=limma.table.8.m[c,]
  #limma.8.m.down.sig=rownames(limma.table.8.m.down[which(limma.table.1.8.down$P.Value < 0.05),])
  #c=which(limma.table.8.m$logFC > 4)
  
  c=which(annotated.limma.8.m$logFC > 2)
  limma.table.8.m.up=annotated.limma.8.m[c,]
  limma.8.m.up.sig=limma.table.8.m.up[which(limma.table.8.m.up$P.Value < 0.05),]
  limma.8.m.up.sig=arrange(limma.8.m.up.sig,desc(limma.8.m.up.sig$logFC))  
  
  #Transcript of Top 20 genes
  interesting_genes.8.m=rownames(limma.8.m.up.sig[1:20,])
  #Name of Top 20 genes
  limma.8.m.up.sig[1:20,]$`ensembl.new$Gene.name`
  
  k = which(rownames(fusion.tra.expression.tra.table.ensembl.table)%in% interesting_genes.8.m)
  View(fusion.tra.expression.tra.table.ensembl.table[k,])
  
  j = which(ensembl.data$Transcript.stable.ID%in%interesting_genes.8.m )
  View(ensembl.data[j,])
```

```{r between m - b stadium}
  c=which(annotated.limma.m.b$logFC > 2)
  limma.table.m.b.up=annotated.limma.m.b[c,]
  limma.m.b.up.sig=limma.table.m.b.up[which(limma.table.m.b.up$P.Value < 0.05),]
  limma.m.b.up.sig=arrange(limma.m.b.up.sig,desc(limma.m.b.up.sig$logFC))  
  interesting_genes.m.b=rownames(limma.m.b.up.sig[1:20,])
  
  k = which(rownames(fusion.tra.expression.tra.table.ensembl.table)%in% interesting_genes.m.b)
  View(fusion.tra.expression.tra.table.ensembl.table[k,])
  
  j = which(ensembl.data$Transcript.stable.ID%in%interesting_genes.m.b)
  View(ensembl.data[j,])
```


### Set enrichment analysis

```{r}
# Convert Ensembl Gene IDs to Entrez IDs because Entrez IDs are needed for SEA-code
# Extract unique transcripts in ensembl data
ensembl.unique = unique(ensembl.data$Transcript.stable.ID)
  
```

```{r Annotate Expression set with Ensembl Gene IDs}
a = which(ensembl.unique %in% rownames(human.vsnrma.df2))
ensembl.new = ensembl.data[a,]
  
b = which(rownames(human.vsnrma.df2) %in% ensembl.unique)
human.vsnrma.df2.new = human.vsnrma.df2[b,]
  
ensembl.new = arrange(ensembl.new,Transcript.stable.ID)
human.vsnrma.df2.new = arrange(human.vsnrma.df2.new, rownames(human.vsnrma.df2.new))
  
annotated.human.vsnrma.df2 = cbind(human.vsnrma.df2.new, ensembl.new$Gene.stable.ID)
  
```

```{r Convert Ensembl Gene IDs to Entrez IDs}

#columns(org.Hs.eg.db) # returns list of available keytypes
annotated.human.vsnrma.df2$entrez = mapIds(org.Hs.eg.db,
                                             keys = annotated.human.vsnrma.df2$`ensembl.new$Gene.stable.ID`, #Column containing Ensembl gene ids
                                             column="ENTREZID",
                                             keytype="ENSEMBL")
```

```{r fit table with annotated expression set}
annotated.fit = lmFit(annotated.human.vsnrma.df2[,1:18], design)
```

```{r Create fit tables for every comparison}
fit.function = function(i){
    fit.table = contrasts.fit(annotated.fit, contrasts = cm[,i])
    fit.table = eBayes(fit.table)
  }
  
fit.1.2 = fit.function(1)
fit.1.4 = fit.function(2)
fit.1.8 = fit.function(3)
fit.1.m = fit.function(4)
fit.1.b = fit.function(5)
fit.2.4 = fit.function(6)
fit.2.8 = fit.function(7)
fit.2.m = fit.function(8)
fit.2.b = fit.function(9)
fit.4.8 = fit.function(10)
fit.4.m = fit.function(11)
fit.4.b = fit.function(12)
fit.8.m = fit.function(13)
fit.8.b = fit.function(14)
fit.m.b = fit.function(15)
```
  
  
```{r create a function to perform SEA on every fit table}
SEA.function = function(x){
    enrich_kegg = kegga(x, geneid = annotated.human.vsnrma.df2$entrez, species = "Hs")
    topKEGG(enrich_kegg)
  }
  
SEA.1.2 = SEA.function(fit.1.2)
# No DE genes
SEA.1.4 = SEA.function(fit.1.4)
SEA.1.8 = SEA.function(fit.1.8)
SEA.1.m = SEA.function(fit.1.m)
SEA.1.b = SEA.function(fit.1.b)
SEA.2.4 = SEA.function(fit.2.4)
SEA.2.8 = SEA.function(fit.2.8)
SEA.2.m = SEA.function(fit.2.m)
SEA.2.b = SEA.function(fit.2.b)
SEA.4.8 = SEA.function(fit.4.8)
SEA.4.m = SEA.function(fit.4.m)
SEA.4.b = SEA.function(fit.4.b)
SEA.8.m = SEA.function(fit.8.m)
SEA.8.b = SEA.function(fit.8.b)
SEA.m.b = SEA.function(fit.m.b)
```

### Venn Diagramm

```{r Venn Diagramm }
##creating demo data
#1.8.tra Upregulated
tra.1.8.up=annotated.limma.1.8.tra[which(annotated.limma.1.8.tra$logFC>0),]$ensembl.transcript
#8.m.tra Upregulated
tra.8.m.up=annotated.limma.8.m.tra[which(annotated.limma.8.m.tra$logFC>0),]$ensembl.transcript
#m.b.tra Upregulated
tra.m.b.up=annotated.limma.m.b.tra[which(annotated.limma.m.b.tra$logFC>0),]$ensembl.transcript

list_venn.up = list(A = tra.1.8.up,B=tra.8.m.up,C=tra.m.b.up)

# Helper function to display Venn diagram.up
display_venn.up <- function(x){
  grid.newpage()
  venn_object <- venn.diagram(x,main=("upregulated TRAS between diffrent stages"), category.names=c("1-cell and 8-cell stage"," 8-cell and morula stage","morula and blastocyst stage"),fill = c("#009E73", "#E69F00", "#56B4E9"), col.main= "red", filename = NULL)
  grid.draw(venn_object)
}

#venn plot up
display_venn(list_venn.up)

Intersect <- function (x) {  
  # Multiple set version of intersect
  # x is a list
  if (length(x) == 1) {
    unlist(x)
  } else if (length(x) == 2) {
    intersect(x[[1]], x[[2]])
  } else if (length(x) > 2){
    intersect(x[[1]], Intersect(x[-1]))
  }
}

intersect.tra.genes.up=Intersect(list_venn.up)

View(fusion.tra.expression.tra.table.ensembl.table[rownames(fusion.tra.expression.tra.table.ensembl.table) %in% intersect.tra.genes.up,])


#1.8.tra downregulated
tra.1.8.down=annotated.limma.1.8.tra[which(annotated.limma.1.8.tra$logFC<0),]$ensembl.transcript
#8.m.tra downregulated
tra.8.m.down=annotated.limma.8.m.tra[which(annotated.limma.8.m.tra$logFC<0),]$ensembl.transcript
#m.b.tra downregulated
tra.m.b.down=annotated.limma.m.b.tra[which(annotated.limma.m.b.tra$logFC<0),]$ensembl.transcript

list_venn.down = list(A = tra.1.8.down,B=tra.8.m.down,C=tra.m.b.down)

# Helper function to display Venn diagram.down
display_venn.down <- function(x){
  grid.newpage()
  venn_object.down <- venn.diagram(x, main=("downregulated TRAS between diffrent stages"), cex.main=2, font.main=4, col.main= "red", category.names=c("1-cell and 8-cell stage"," 8-cell and morula stage","morula and blastocyst stage"),fill = c("#009E73", "#E69F00", "#56B4E9"), filename = NULL)
  grid.draw(venn_object.down)
}
display_venn.down(list_venn.down)

intersect.tra.genes.down=Intersect(list_venn.down)

View(fusion.tra.expression.tra.table.ensembl.table[rownames(fusion.tra.expression.tra.table.ensembl.table) %in% intersect.tra.genes.down,])
```

