---
title: "__The role of tissue-restricted antigens in the early embryonic development
 __"
date: "18.07.2022"
output:
  pdf_document: default
  html_document:
    df_print: paged
header-includes:
    \usepackage{float}
    \floatplacement{figure}{H}
    \usepackage{caption}
    \captionsetup[figure]{font=footnotesize}
---
\thispagestyle{empty}
\hrule
\vspace{0.3cm}
\begin{center}
	\vspace{1cm}
  \large
	\begin{tabular}[c]{l}
	 \\

Clara Certa, Yaxin Chen, Linda Kaupp, Alewtina Towara \\
\\
	 \\
	Supervisor:
	Dr. Maria Dinkelacker, Dr. Carl Herrmann \\
	 \\
	Tutor:
	Ian Dirk Fichter \\
	 \\
	 \\
	Data analysis for students of Molecular Biotechnology \\
	\\
	Heidelberg University 
	\vspace{0.3cm} \\
	\end{tabular}
	\end{center}

---

\pagenumbering{gobble}
\pagebreak
\tableofcontents
\pagebreak
\pagenumbering{arabic}


```{r setup, include=FALSE}
knitr::opts_chunk$set(include=FALSE, echo = FALSE, fig.align = 'centre', message=FALSE,warning = FALSE, cache = TRUE)
knitr::opts_knit$set(root.dir = "~//Documents//GitHub//2022-topic-04-team-03")
```

```{r}
projectPath <- dirname(rstudioapi::getSourceEditorContext()$path)
```


\newpage

# 1.Abstract
Research, especially in the field of early embryogenesis is important in many ways. It is not only required for better understanding of genetic defects of the embryos in an early stage it is also important for improvement of assisted human conception. In this report we investigated gene expression data of the early stages of embryogenesis up to the formation of the blastocyst. As the formation of tissues only occurs after the stages displayed by our dataset, we expected to not observe any tissue specific development and therefore no differentially expressed tissue-restricted antigens (TRAs) during this period. However, we found numerous differentially expressed tissue-restricted antigens that display the developmental processes in early embryogenesis accurately. For instance, the inquired genes reflect the activation of transcription of the embryonic genome as well as the switch from anaerobic to aerobic energy metabolism in the early embryo.
Furthermore, we found three differentially expressed chemokines that play a role in ??



# 2.Introduction
## 2.1 Earyl embryogenesis
After the fertilization the human embryo begins to cleave every 24 hours. The individual cells can be detected until the 16-cell stage. Till then every single cell is able to form a whole fetus as well as the fetal placenta. After the 16-cell stage an event called compaction occurs. The individual cells, or blastomeres, become flat and thereby maximize the cell-cell contacts to one another. The result of compaction is the morula stage. The next step is the formation of the blastocyst which is the first differentiative step in early embryo development. The outer cells form an epithelial layer around the embryo called the trophectoderm. The trophectoderm is later important for implantation in the uterine endometrium. As soon as the epithelial layer is built those cells start the ingestion of fluid which leads to the formation of the blastocoel cavity. The blastocoel is additionally containing a small number of undefined cells which will later form the actual embryo. Those cells are referred to as the inner cell mass (Leese *et al*.,1993).
Interestingly the activation of the embryo genome is thought to only start after the 4-cell stage. Prior the development is driven by maternal stores of RNA which were passed during oogenesis. 
The metabolic activity itself only starts rising at the late morula stage, after a process called the maternal to zygotic transition (MZT) (Schier, 2007). The two reasons for that are the increased energy demand caused by protein biosynthesis as the number of dividing cells goes up and the ion pumping into the blastocoel cavity. 
During this phase the energy metabolism of the embryo changes from using aerobic substrates such as pyruvate, lactate, and glutamine to using glucose (Leese *et al*., 1993).
This switch results in a 2.7-fold rise of oxygen consumption in the trophectoderm where the oxidative phosphorylation takes place. As a consequence, the production of harmful reactive oxygen species (ROS) increases. The embryo therefore has to invent a system to prevent high amounts of ROS and at the same time prepare repair mechanisms for fixing the damage (Kaneko *et al*., 2013).

## 2.2 Chemokines
The secretion of signal proteins known as chemokines plays an important cause different biological processes like the movement of epithelial cells or leukocytes (Raman et al., 2011). Those effects also occur during pregnancy and it has been shown that chemokines play an essential role in the implantation of the embryo and embryonic development (Hannan et al., 2007)
\newpage

## 2.3 Tissue-restricted antigens
Our immune system has the important role of protecting our body from exogenous pathogens. The immune system consists of two parts, the innate immune system and the adaptive immune system. The innate immune system reacts very quickly to foreign substances but is restricted to recognizing conserved structures like bacterial cell wall components. Whereas the adaptive immune system develops after birth and is able to target a great variety of antigens. An important part of the adaptive immune response are T cells which recognize antigens with specific T cell receptors (TCRs). 
In this process it is crucial that the immune system solely attacks foreign pathogens and not the body’s own tissues. To gain self-tolerance a selection process takes place in the thymus where T cells are screened for T cell receptors that can recognize antigens but not self-antigens. First, the T cells undergo positive selection. In this step only those T cells that can recognize self-major histocompatibility complex (MHC) proteins are selected to survive. Thereafter, a negative selection process takes places which prevents the emergence of autoreactive T cells. This is based on triggering of apoptosis in T cells with TCRs that are capable of recognizing self-antigens in the context of self MHC. These self-antigens are referred to as tissue-restricted antigens (TRAs) and are presented by medullary thymic epithelial cells (mTECs) (Murphy, 2012).
TRAs are genes which are expressed considerably higher in certain tissues compared to other tissues. Thus, TRAs are characteristic for the tissues they belong to. However, in mTECs, TRAs are expressed promiscuously to ensure that all tissues are represented in the thymus. Hence, every autoreactive T cell is sorted out (Dinkelacker et al., 2007).

## 2.4 Aim of this project
This project aims to investigate the expression of TRAs in the early embryonic development. Tissue-specifity plays an important role in embryonic development as embryonic stem cells differentiate step by step to eventually form organs. This development could be monitored by analyzing gene expression data of different stages in embryogenesis. 
In this project, we are investigating the preimplantation stages of embryonic development up to the formation of the blastocyst. In the course of our work, we will determine the differentially expressed genes between the first few stages of embryonic development and extract the differentially expressed TRAs for further analysis. 
Chemokines??
\newpage

# 3.Materials and methods
## 3.1 Datasets
In order to study the cellular and tissue development during human early embryogenesis we analyzed a microarray dataset published by Xie et al.in 2010. We examined the amount of differentially expressed tissue-restricted antigens (TRAs) in our dataset by comparing it with the TRA dataset provided by Dr. Dinkelacker in 2019.

### 3.1.1 Microarray dataset
The microarray dataset by Xie et all. contained three datasets one of bovine, one of mouse and one of human. We decided to focus our project on the human dataset. The data was generated by RNA extraction, amplification, and hybridisation onto Affymetrix microarrays. For early human embryogenesis they prepared three replicates at one cell stage, two cell stage, four cell stage, eight cell stage, morula, and blastocyst each. Each replicate included 95,659 transcripts.

### 3.1.2 TRA dataset
As a reference TRA dataset we chose the GTEX RNAseq human dataset provided by Dr. Dinkelacker in 2019. It consisted of 60,131 transcripts. We used the TRA dataset and ensembl gene information to annotate our raw data.

## 3.2 Packages used in R
During of the analysis a number of packages was used which are listed in Table 1. The code was written using the 4.2.0 Version of R.
\begin{table}[htb]
\caption{List of all R and Bioconducter packages used in our project.}
\resizebox{\textwidth}{!}{
\begin{tabular}{lllll}
\hline
affy 1.74.0 & vsn 3.64.2 & ggplot2 3.3.6 & AnnotationDbi 1.58.0 & VennDiagram 1.7.3 \\ 
vsn 3.64.0 & limma 3.52.2  & dplyr 1.0.9 & tidyverse 1.3.1 & hexbin 1.28.2 \\ 
hgu133plus2hsenstcdf 25.0.0  & cluster 2.1.3 & factoextra 1.0.7 & org.Hs.eg.d 3.15.0 & go.db 3.15.0\\
hgu133plus2hsenstprobe 25.0.0 & pheatmap 1.0.12 & imager 0.42.13 & tinytex 0.40 & RColorBrewer 1.1-3 \\
clusterProfiler 4.4.4 & enrichplot 1.16.1 & \\
\hline
\end{tabular}}
\end{table}

```{r load packages, eval=TRUE, include=FALSE, comment=FALSE}
#load library
library(affy)
library(vsn)
library(AnnotationDbi)
library(hgu133plus2hsenstcdf)
library(hgu133plus2hsenstprobe)
library(limma)
library(pheatmap)
library(dplyr)
library(tidyverse)
library(hexbin)
library(cluster)
library(factoextra)
library(EnhancedVolcano)
library(org.Hs.eg.db)
library(RColorBrewer)
library(VennDiagram)
library(GO.db)
library(tinytex)
library(clusterProfiler)
library(enrichplot)
library(ggplot2)
library(imager)
```



## 3.3 Quality control and normalization of the data

Before analyzing the microarrays, it is important to determine the quality of the used raw data. This is done by the procedure called Quality Control (QC) where potential mistakes in the experiment can be identified.  

```{r set working directory or read in .CEL files}
# setting working directory and loading
setwd("~//Documents//GitHub//2022-topic-04-team-03//Data//rawData")

data.human=ReadAffy()
data.human@cdfName <- "HGU133Plus2_Hs_ENST"

```
\newpage
### 3.3.1 Single chip control
The first step of QC is the single-chip analysis where major quality problems like fingerprints, local irregularities, imprints of pipette tips and extreme light intensities can be detected by a visual inspection of the microarray images. After looking at our chips we presupposed that the GSM456645 chip could be of bad quality due to a smudge on the bottom edge.

```{r chips pictures, eval =FALSE}
for (i in 1:18){
   image(data.human[,i], col = rainbow (100, start = 0, end = 0.75)[100:1],main=substr(colnames(data.human)[i],1,9)) }
```

```{r only chip GSM456645,  include=TRUE, fig.cap="Single chip image of chip GSM456645",fig.height=3.4, fig.width=3.2,fig.align="center"}
image(data.human[,3], col = rainbow (100, start = 0, end = 0.75)[100:1],main=substr(colnames(data.human)[3],1,9))
```
 
### 3.3.2 RNA degradation plot
In the next step we analysed the RNA degradation that is an indicator for a bad quality RNA. Due to the start of the RNA degradation at the 5´ sequence end the intestines of probes at the 5´ end must be lower than at the 3´ end. In our RNA degradation plot, where the mean intensity for each probe set is plotted versus the probe number, no high slope and disagreements between the arrays were detected.

```{r RNA Degradation Plot, eval =FALSE, include=TRUE}
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//RNAdegrad")
rnadeg.raw = AffyRNAdeg(data.human)

plotAffyRNAdeg(rnadeg.raw, col=rainbow(35))
title(sub="human embroyogenesis rawdata")
dev.copy2pdf(file="RNAdegrad_plot.pdf", width = 12.5, height = 20)
```


```{r RNA Degradation Plot shifted, eval =FALSE, include=TRUE, fig.cap="Figure 2: RNA degradation plot"}
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//RNAdegrad")
plotAffyRNAdeg(rnadeg.raw, col=rainbow(35), transform= "shift.only")
title(sub="human embryogenesis rawdata - shifted")

dev.copy2pdf(file="RNAdegrad_plot_shifted.pdf", width = 12.5, height = 20)
```

### 3.3.3 Normalization
By performing the variance stabilization normalization (vsnrma) (Huber et al.) the data is corrected for biases from non-biological sources, like labelling efficiency and scanner setup. We also measured the quality of the variance stabilization by plotting the row standard deviation versus row means. 
```{r normalization, echo=FALSE, message=FALSE, comment=FALSE, results="hide", highlight=TRUE}
#Normalization

human.vsnrma <- vsnrma(data.human)
setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")
colnames(human.vsnrma)= substr(colnames(human.vsnrma), 1,9)
expression.data <- exprs(human.vsnrma)
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
```

The red line in this meanSdplot is not an approximately horizontal as it should be. This could be caused by the standard deviation at high expression levels being higher than at low levels even if the relative deviation from mean is identical.

```{r meanSdPlot, include=TRUE, fig.cap="MeanSd plot", fig.height=2.7,fig.width=4.5,fig.align='center'}
meanSdPlot(human.vsnrma)

#setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//MeanSdPlot")
#dev.copy2eps(file="meanSdPlot_human_vsnrma_normalized.eps")
###The line is not approximately horizontal, but like in project proposal said, so there shouldn't be a variance-mean dependence 
```



Different plots were reviewed in order to determine the degree of the systematic variation in the raw data. For both, the pre-normalized data and the normalized data, density plots and box plots were generated.

<!-- #### 3.3.3.1 Boxplots -->
The box plots visualize the variation of the gene expression for each chip. After the normalization no boxes were significantly elevated or more spread out than other arrays.
```{r boxplots, eval =FALSE, include=TRUE, fig.cap= "Figure 3: Normalization of raw data"}
names = c("1 cell stage, rep 1", "1 cell stage, rep 2", "1 cell stage, rep 3",
          "2 cell stage, rep 1", "2 cell stage, rep 2", "2 cell stage, rep 3",
          "4 cell stage, rep 1", "4 cell stage, rep 2", "4 cell stage, rep 3",
          "8 cell stage, rep 1", "8 cell stage, rep 2", "8 cell stage, rep 3",
          "morula , rep 1", "morula , rep 2", "morula , rep 3",
          "blastocyst , rep 1", "blastocyst , rep 2", "blastocyst , rep 3")


# Before normalization:
par(las=2)
mmi=c(1,0.7,1.0477939,0.5366749)
par(mai=mmi)
boxplot(data.human, col= rainbow(35), cex.axis=0.5, main="Gene expression in human embroyogenesis data before normalization", names =names,las=2)
title(ylab="relative expression", line=2)
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//Boxplot")
dev.copy2eps(file="boxplot_rawdata_18290.eps")

# After Normalization:
boxplot(exprs(human.vsnrma), col= rainbow(35), cex.axis=0.5, main="Gene expression in human embroyogenesis data after normalization",names =names,las=2)
title(ylab="relative expression", line=2)
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//Boxplot")
dev.copy2eps(file="boxplot_normalizeddata_18290.eps")
```




<!-- #### 3.3.3.2 Density plots -->
In the density plots for each chip the density function is plotted versus log expression. After the vsnrma normalization the distribution of all curves in the plot showed a vicinity. 
```{r density plots, eval =FALSE, include = TRUE, fig.cap= "Figure 4: Density plot"}
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//Density")
# Before Normalization:
hist(data.human, col=rainbow(35), main="Density function of log Intensity of human embryogenesis raw data")


dev.copy2pdf(file="Histogram_rawdata_39897.pdf", width = 10, height = 7)

# After Normalization:
eset=exprs(human.vsnrma)

plot(density(eset[,1]), type="n", xlab="log Intensity", ylim=c(0,1), main="Density function of log Intensity of human embryogenesis normalized")
for (i in 1:ncol(eset)) {
  lines(density(eset[,i]), col=rainbow(35)[i])
}

dev.copy2pdf(file="Histogram_NormalizedData_39897.pdf", width = 10, height = 7)
```



<!-- #### 3.3.3.3 Scatter plots -->
For the control of the normalization the scatter plots, that compare the log fold intensity change between an array and a reference array, were made.

```{r scatter plot, eval =FALSE}
expression.data <- exprs(human.vsnrma)

for(i in 1:17){
  plot(expression.data[,c(i,i+1)], pch=".", cex=2)
  abline(0, 1, col="red")               # 45 degree dividing line
  
  title(main = paste("Scatterplot of probe", 
                     substr(colnames(human.vsnrma)[i], 1, nchar(colnames(human.vsnrma)[i])), "and", 
                     substr(colnames(human.vsnrma)[i+1], 1, nchar(colnames(human.vsnrma)[i+1])), 
                     sep=" ", collapse = NULL))
  
  file.name <- paste("~//documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//Scatterplot", 
                     as.character(substr(colnames(human.vsnrma)[i], 1, nchar(colnames(human.vsnrma)[i]))), "_",
                     as.character(substr(colnames(human.vsnrma)[i+1], 1, nchar(colnames(human.vsnrma)[i+1]))),
                     ".pdf", sep="")

  dev.copy2pdf(file = file.name)
}
```

As a consequence of our Quality Control we decided to keep all chips for further analysis, because no serious outliers after the normalization were found.

<!-- Data Clean-up -->

```{r data clean-up}
# Create a data frame out of the expression data from the normalized data
human.vsnrma.df = data.frame(exprs(human.vsnrma))

# Are there any NAs in our dataset?
sum(apply(human.vsnrma.df,1,function(x){sum(is.na(x))}))
#> No

```

## 3.4 Principal component analysis
Principal Component Analysis (PCA) is an important method to reduce the dimensionality of a data set by decreasing its number of variables, which do not contribute to the variance of the data. Principal components (PC) are calculated, which are linear combinations of variables that explain a portion of variance among the total variance in the data. By looking at the PCs that explain an accumulative variance of over 90% and plotting them against each other, a better understanding of the data can be achieved, for example, by discovering potential clusters of the data points.

Quelllee!!!!!!!!!!!!!!

## 3.5 K-means clustering
K-means is a clustering method for which one first must indicate the number of clusters (k). Later the ideal number can be detected by an elbow-plot or silhouette-plot. The algorithm starts by choosing random centroids of the k clusters. Each object is now assigned to the cluster for which the Euclidean distance between centroid and object is the smallest. In the next step the centroids get updated by calculating the mean of each cluster containing the new objects. Those steps are repeated till the cluster assignments of the objects stop changing (Bortz,2010).

## 3.6 Limma analysis
For anaylsis of differential gene expression we conducted limma analysis. Therefore, the 	“Limma” package was employed, which facilitates statistical testing of expression level 		changes of thousands of genes at once. For that, “Limma” uses linear models to calculate 	the test statistic between sample group expression levels (Ritchie et al.,2015).

## 3.7 Gene set enrichment analysis
Gene set enrichment analysis (GSEA) is a method to identify groups of genes in a dataset that are overrepresented more than expected in any known set of genes. These gene sets can be groups of genes that share a common biological function, chromosomal location, or regulation. GSEA uses statistical approaches to find groups of genes that are significantly enriched or depleted. The goal of this method is to better understand the biological function of groups of genes in a data set (Subramanian et al.,2005).  

# 4. Results
<!-- ## Annotation of dataset using TRAs -->


```{r Ensembl data}
# read in ensembl table with following attributes: "Gene stable ID", 
#"Gene stable ID version", "Transcript stable ID", "Transscript stable ID version", 
#"Gene.name", "Transcript name", "Chromosome.scaffold.name", "Gene.description", "HGNC.symbol"

setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")

ensembl.data <- read.csv("ensembl.human.txt")
dim(ensembl.data)
##268341      9

# create variables that contain gene ID, transcript ID, 
#the chromosome name and the gene symbol
ensembl.genes = ensembl.data[,1]
ensembl.transcripts = ensembl.data[,3]
ensembl.chromosome = ensembl.data[,7]
ensembl.symbols = ensembl.data[,9]
```



```{r clean human.vsnrma.df}
#rename the colnames to the stages
names = c("1 cell stage, rep 1", "1 cell stage, rep 2", "1 cell stage, rep 3",
          "2 cell stage, rep 1", "2 cell stage, rep 2", "2 cell stage, rep 3",
          "4 cell stage, rep 1", "4 cell stage, rep 2", "4 cell stage, rep 3",
          "8 cell stage, rep 1", "8 cell stage, rep 2", "8 cell stage, rep 3",
          "morula stage, rep 1", "morula stage, rep 2", "morula stage, rep 3",
          "blastocyst stage, rep 1", "blastocyst stage, rep 2", "blastocyst stage, rep 3")
colnames(human.vsnrma.df) = names

# Check dimensions
dim(human.vsnrma.df)
#95,721    18

# exclude Affymetrix control genes which begin with "AFFX"
human.vsnrma.df2 = human.vsnrma.df[63:95721,]
dim(human.vsnrma.df2)
# 95,659    18

# remove ".xx_at" suffix
rownames(human.vsnrma.df2) = substr(rownames(human.vsnrma.df2), 1,15)
```


```{r TRA data}
# read in TRA data for human
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")

tra.data <- read.table("tra.2017.human.gtex.5x.table.tsv",header=TRUE)
dim(tra.data)
#60131    10

# find the index of rownames of our dataset that contain transcript Ids from the tra dataset
# We extracted the TRA genes out of the three tables(human.vsnrma, tra, ensembl)
#TRAs in our dataset
j = which(rownames(human.vsnrma.df2) %in% tra.data$ensembl.transcript) 
tra.extracted = rownames(human.vsnrma.df2)[j] #24,783
human.vsnrma.only.tra = human.vsnrma.df2[j,]
dim(human.vsnrma.only.tra)
#24783 18 

#exclude TRAs which are not expressed in our dataset
k = which(tra.data$ensembl.transcript %in% tra.extracted)
tra.chips = tra.data[k,] #24,783 
dim(tra.chips)
#24783 10
```



```{r Extract TRAs in Ensembl }
#we want to now extract the genes from Ensembl table that are present in the TRAs expressed in our chips
c = which(ensembl.transcripts %in% tra.extracted)
ensembl.only.tra = ensembl.data[c,]
dim(ensembl.only.tra)
#24623 9
```

<!-- There are 160 genes which are in the TRA table but not in Ensembl table. We downloaded and checked the Ensembl table again, which did not help. -->


```{r TRA Annotation}
#reorder the rows of ensembl.only.tra and tra.chips
tra.chips = arrange(tra.chips,ensembl.transcript)
ensembl.only.tra = arrange(ensembl.only.tra,Transcript.stable.ID)

#fusion of two tables: tra.chips and human.vsnrma.only.tra
fusion.tra.chips = cbind(human.vsnrma.only.tra,tra.chips$ensembl.gene,tra.chips$tiss.number,tra.chips$tissues,tra.chips$max.tissue)
colnames(fusion.tra.chips)[19:22]<-c("ensembl.gene","tissue.number","tissues","max.tissue")

# fusion of fusion.tra.chips and ensembl.only.tra (aware that ensembl.only.tra table contains less rows than the other two tables due to the 160 missing genes)
p = which(rownames(fusion.tra.chips) %in% ensembl.transcripts)
tra.ensembl = fusion.tra.chips[p,]
annotated.tra = cbind(tra.ensembl,ensembl.only.tra$Gene.name,ensembl.only.tra$Gene.description,ensembl.only.tra$Chromosome.scaffold.name)

#rename colnames
colnames(annotated.tra)[23:25]<-c("Gene.name","Gene.description","Chromosome.scaffold.name")


setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
write.csv(annotated.tra, file="annotated.tra.csv")

```

## 4.1 Tissue distribution plot
We wanted to investigate our hypothesis whether or not there are even differential expressed TRAs in our dataset by visualizing the distribution of the TRAs differential expressed in our chips according to their max tissue, using a barplot.

```{r Tissue distribution image, include = TRUE,fig.cap="Frequency of TRAs differentially expressed between 1-cell stage and 8-cell stage", fig.height=5,fig.width=5}
library(imager)
img<-load.image('tissue.distribution.arranged.barplot.jpeg') 
plot(img, axes=F)
```


It seems like even though there is no tissue development in the observed stages certain TRAs are still differentially expressed.


## 4.2 Principal component analysis
In order to verify that the variance in our dataset results from the biological background more than from the technical process we performed a principal component analysis. Moreover, we thereby achieved a greater insight in possible interesting stages to examine. 

```{r PCA}
#conducting PCA
topVar = apply(human.vsnrma.df2, 1, var)
q75 = quantile(topVar, probs = 0.75)
i.topvar = which(topVar >= q75)
human.vsnrma.df2.topVar = human.vsnrma.df2[i.topvar,]

pca = prcomp(t(human.vsnrma.df2.topVar), center = T, scale. = T)


```

```{r show how much variance each PC explains}
#show how much variance each PC explains

variance = (pca$sdev)^2
prop.variance = variance/sum(variance)

# Plot cumulative proportion of variance explained
plot(cumsum(prop.variance), xlab = "Principal component",
     ylab = "Cumulative proportion of variance explained",cex.lab=0.7,
     ylim = c(0, 1), type = "b")

abline(h=0.9)
#4 PCs are enough to show more than 90%
```



```{r Plot PC1 and PC2, include=TRUE, fig.cap="Principal component analysis",fig.height=3.5, fig.width=7, fig.align='center'}
color1 = c(rep("red",3),rep("orange",3),rep("yellow",3),rep("green",3),rep("blue",3),rep("purple",3))

par(mar=c(5, 4, 4, 10), xpd=TRUE)

plot(pca$x[,1], pca$x[,2],col=color1, pch=19,xlab='PC1',ylab='PC2',main="Principal component analysis")

color2= c("red","orange","yellow","green","blue","purple")

legend(x = 170,y=130,legend = c("1-cell stage", "2-cell stage","4-cell stage", "8-cell stage", "Morula stage","Blastocyst stage"), fill= color2,col = color2)


```

<!-- We cannot observe any outlier replicate and we can see four cluster (Cluster 1:1-cell,2-cell and 4-cell stage, Cluster 2: 8-cell stage, Cluster 3:Morula, Cluster 4:Blastocyst) -->

```{r plots PC2/3 and PC3/4, eval=FALSE}
#Plot PC2 and PC3

par(mar=c(5, 4, 4, 10), xpd=TRUE)

plot(pca$x[,2], pca$x[,3],col=color1, pch=19,xlab='PC2',ylab='PC3',main="Principal component analysis")

legend(x = 140,y=92,legend = c("1-cell stage", "2-cell stage","4-cell stage", "8-cell stage", "Morula stage","Blastocyst stage"), fill= color2,col = color2)

#Plot PC3 and PC4
par(mar=c(5, 4, 4, 10), xpd=TRUE)

plot(pca$x[,3], pca$x[,4],col=color1, pch=19,xlab='PC3',ylab='PC4',main="Principal component analysis")

color2= c("red","orange","yellow","green","blue","purple")

legend(x = 100,y=48,legend = c("1-cell stage", "2-cell stage","4-cell stage", "8-cell stage", "Morula stage","Blastocyst stage"), fill= color2,col = color2)


```
We grouped our replicates using k-means clustering on the first 4 principal components which eyplain 90% of the variance. By performing a silhouette plot we indicated four as the number of clusters.The clusters were composed as follows:Cluster 1(1 cell,2 cell and 4 cell stage), Cluster 2(8 cell stage), Cluster 3 (morula stage) and Cluster 4 (blastocyst stage). Due to this result, we focused our project on investigating the difference in gene expression between 1-cell stage and 8-cell stage, 8-cell stage and morula and morula and blastocyst. 


```{r run models with varying value of k}
#Use map_dbl to run many models with varying value of k (centers)
tot_withinss <- map_dbl(1:10,  function(k){
  model <- kmeans(x = pca$x[,1:4], centers = k,nstart=30)
  model$tot.withinss
})
```

```{r data frame with k & tot_withinns}
# Generate a data frame containing both k and tot_withinss
elbow_df <- data.frame(
  k = 1:10 ,
  tot_withinss = tot_withinss
)
```

```{r plot with elbow plot, echo=TRUE}
ggplot(elbow_df, aes(x = k, y =tot_withinss)) +
  geom_line() +
  scale_x_continuous(breaks = 1:10) + labs(y= "Total within sum of squares")+ ggtitle("Elbow plot")

#seems like k=2 is the best
```



```{r show cluster, echo=TRUE}
model <- kmeans(x = pca$x[,1:4], centers = 2,nstart=30)
b<- as.data.frame(model$cluster)
colnames(b) <- "Cluster"
b

#Elbow plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)
```






```{r Use map_dbl to run many models with varying value of k}
#Plot the Siloutte plot
sil_width <- map_dbl(2:10,  function(k){
  model <- pam(pca$x[,1:4], k = k)
  model$silinfo$avg.width
})
```

```{r Silhouette plot, fig.cap= "Silhouette plot",fig.height=3, fig.width=3, fig.align='left'}
# Generate a data frame containing both k and sil_width
sil_df <- data.frame(
  k = 2:10,
  sil_width = sil_width
)

# Plot the relationship between k and sil_width
ggplot(sil_df, aes(x = k, y = sil_width)) +
  geom_line() +
  scale_x_continuous(breaks = 2:10) + labs(y= "Silhouette Width")+ ggtitle("Silhouette plot")
```


```{r Cluster silhouette plot,  fig.subcap= "Table 2: Clusters determined by silhouette plot",fig.align='right'}
# Generate a k-means model using the pam() function with a k = 4
pam_k4 <- pam(pca$x[,1:4],k=4)

# Show the clustering
a <-as.data.frame(pam_k4$clustering)
colnames(a)= "Cluster"

knitr::kable(a,format = "simple",caption ="Clusters silhouette plot")

```


<!-- Silouette Plot suggests 4 Clusters: Cluster 1:1-cell,2-cell and 4-cell stage, Cluster 2: 8-cell stage, Cluster 3:Morula, Cluster 4:Blastocyst) -->


<!-- Using Kmeans based on the original expression data (human.vsnrma.df2) to cluster and compare the results with the Kmeans results based on 4PCs -->

```{r kmeans on whole data}
# Use map_dbl to run many models with varying value of k (centers)
tot_withinss <- map_dbl(1:10,  function(k){
    model <- kmeans(x = t(human.vsnrma.df2), centers = k,nstart=30)
    model$tot.withinss
})
```

```{r Elbow plot}
# Generate a data frame containing both k and tot_withinss
elbow_df <- data.frame(
    k = 1:10 ,
    tot_withinss = tot_withinss
)

# Plot the elbow plot
ggplot(elbow_df, aes(x = k, y =tot_withinss)) +
    geom_line() +
    scale_x_continuous(breaks = 1:10)+ labs(y= "Total within sum of squares")+ ggtitle("Elbow plot")
#seems like k=2 is the best
```



```{r}
#show cluster
model <- kmeans(x = t(human.vsnrma.df2), centers = 2,nstart=30)
c<-as.data.frame(model$cluster)
colnames(c)="Cluster"
c

#Elbow plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)
```






```{r Silhoutte plot whole data}
#Plot the Silhoutte plot
# Use map_dbl to run many models with varying value of k
sil_width <- map_dbl(2:10,  function(k){
    model <- pam(t(human.vsnrma.df2), k = k)
    model$silinfo$avg.width
})

# Generate a data frame containing both k and sil_width
sil_df <- data.frame(
    k = 2:10,
    sil_width = sil_width
)

# Plot the relationship between k and sil_width
ggplot(sil_df, aes(x = k, y = sil_width)) +
    geom_line() +
    scale_x_continuous(breaks = 2:10)+ labs(y= "Silhouette Width")+ ggtitle("Silhouette plot")

#it seems like that k=2 is a good choice
```



```{r}
# Generate a k-means model using the pam() function with a k = 4
pam_k2 <- pam(t(human.vsnrma.df2),k=2)
  
# Show the clustering
d<- as.data.frame(pam_k2$clustering)
colnames(d) <- "Cluster"
d

#Silouette plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)
```





## 4.3 Limma analysis

We performed limma analysis using the normalized expression data to discover the differentially expressed genes (DEGs) between every combination of stages. We filtered the resulting DEGs for a p-value of 0.01 and annotated the genes with ensembl Biomart and the TRA data. Based on our relevant clusters found by K-means clustering, we solely focused on the DEGs between stages 1-cell to 8-cell, 8-cell to morula and morula to blastocyst for further analysis. For the comparison of the 1-cell to 8-cell stage, we found xx DEGs, for 8-cell stage to morula stage there were xx DEGs and for morula stage to blastocyst we got xx DEGs.

```{r limma Analysis}

# Define the different stages of the chips in a vector
stage = factor(x= c(1,1,1,2,2,2,4,4,4,8,8,8,"morula","morula","morula", "blastocyst", "blastocyst", "blastocyst"),levels= c(1,2,4,8,"morula","blastocyst"))

# Create a design matrix with the stages
design = model.matrix(~0 + stage)
            
# Create a contrast matrix which compares all the stages with one another 
cm = makeContrasts(stage1.2 = stage2-stage1, stage1.4 = stage4-stage1, stage1.8 = stage8-stage1, stage1.morula = stagemorula-stage1, stage1.blastocyst =stageblastocyst-stage1,
                   stage2.4 = stage4-stage2, stage2.8 = stage8-stage2, stage2.morula = stagemorula-stage2,  stage2.blastocyst = stageblastocyst-stage2,
                   stage4.8 = stage8-stage4, stage4.morula = stagemorula-stage4,  stage4.blastocyst = stageblastocyst-stage4,
                   stage8.morula = stagemorula-stage8, stage8.blastocyst = stageblastocyst-stage8,
                   stagemorula.blastocyst =stageblastocyst -stagemorula, 
                   levels = design)

# Fit the coefficients of the model
fit = lmFit(human.vsnrma.df2, design)
fit2 = contrasts.fit(fit, contrasts = cm)

# Calculate the t-statistics
fit2 = eBayes(fit2)

# Number of differentially expressed genes (p-value = 0.05)
results = decideTests(fit2, p.value = 0.05)
summary(results)

# Number of differentially expressed genes (p-value = 0.01)
results2 = decideTests(fit2, p.value = 0.01)
summary(results2)
```
<!-- *creating fit tables between every two stages* -->

```{r fitFunction}
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")

fit.function = function(i){
  fit.table = contrasts.fit(fit, contrasts = cm[,i])
  fit.table = eBayes(fit.table)
}
```

```{r limma fit tables}
fit.1.2 = fit.function(1)
fit.1.4 = fit.function(2)
fit.1.8 = fit.function(3)
fit.1.m = fit.function(4)
fit.1.b = fit.function(5)
fit.2.4 = fit.function(6)
fit.2.8 = fit.function(7)
fit.2.m = fit.function(8)
fit.2.b = fit.function(9)
fit.4.8 = fit.function(10)
fit.4.m = fit.function(11)
fit.4.b = fit.function(12)
fit.8.m = fit.function(13)
fit.8.b = fit.function(14)
fit.m.b = fit.function(15)
```

<!-- *using toptable function to create limma table (only significant DEGs!) for every two stages* -->

```{r topTable function}
top.table = function(i){
  fit.table = contrasts.fit(fit, contrasts = cm[,i])
  fit.table = eBayes(fit.table)
  
  pvalue01 = sum(p.adjust(fit.table$p.value, "BH") < 0.01)
  topTable(fit.table, number = pvalue01)
}
```

```{r limma topTables}
limma.table.1.2 = top.table(1)
limma.table.1.4 = top.table(2)
limma.table.1.8 = top.table(3)
limma.table.1.m = top.table(4)
limma.table.1.b = top.table(5)
limma.table.2.4 = top.table(6)
limma.table.2.8 = top.table(7)
limma.table.2.m = top.table(8)
limma.table.2.b = top.table(9)
limma.table.4.8 = top.table(10)
limma.table.4.m = top.table(11)
limma.table.4.b = top.table(12)
limma.table.8.m = top.table(13)
limma.table.8.b = top.table(14)
limma.table.m.b = top.table(15)
```

<!-- *we annotate the limma table (only significant DEGs) with ensembl* -->

```{r annotation function}
limma.annotation = function(x){
a = which(ensembl.transcripts %in% rownames(x))
ensembl.new = ensembl.data[a,]
  
b = which(rownames(x) %in% ensembl.new$Transcript.stable.ID)
x.new = x[b,]
  
ensembl.new = arrange(ensembl.new,Transcript.stable.ID)
x.new = arrange(x.new, rownames(x.new))
  
annotated.x = cbind(x.new, ensembl.new$Gene.name, ensembl.new$Gene.description,ensembl.new$Chromosome.scaffold.name, ensembl.new$Gene.stable.ID)
colnames(annotated.x)[7:10]<-c("gene.name","gene.description","Chromosome.scaffold.name", "Gene.stable.ID")
annotated.x = arrange(annotated.x, adj.P.Val)
  return(annotated.x)
}
```

```{r limma annotation}
annotated.limma.1.8 = limma.annotation(limma.table.1.8)
annotated.limma.1.m = limma.annotation(limma.table.1.m)
annotated.limma.1.b = limma.annotation(limma.table.1.b)
#annotated.limma.2.4 = limma.annotation(limma.table.2.4)
annotated.limma.2.8 = limma.annotation(limma.table.2.8)
annotated.limma.2.m = limma.annotation(limma.table.2.m)
annotated.limma.2.b = limma.annotation(limma.table.2.b)
annotated.limma.4.8 = limma.annotation(limma.table.4.8)
annotated.limma.4.m = limma.annotation(limma.table.4.m)
annotated.limma.4.b = limma.annotation(limma.table.4.b)
annotated.limma.8.m = limma.annotation(limma.table.8.m)
annotated.limma.8.b = limma.annotation(limma.table.8.b)
annotated.limma.m.b = limma.annotation(limma.table.m.b)
```

<!-- *we extract the TRAs from ensembl annotated limma table (only significant DEGs) and annotate these significant differentially expressed TRAs with TRA information* -->

```{r limma TRAs}
#annotated limma tables mit TRAs annotaten
#1.4 und 1.2 sind 00
nr<-c(1,9,10)

annotated.limma.1.8$ensembl.transcript = rownames(annotated.limma.1.8)
annotated.limma.1.8.tra <- merge(annotated.limma.1.8, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.1.m$ensembl.transcript = rownames(annotated.limma.1.m)
annotated.limma.1.m.tra <- merge(annotated.limma.1.m, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.1.b$ensembl.transcript = rownames(annotated.limma.1.b)
annotated.limma.1.b.tra <- merge(annotated.limma.1.b, tra.data[,nr], by = 'ensembl.transcript')
#2.4. ist 00
annotated.limma.2.8$ensembl.transcript = rownames(annotated.limma.2.8)
annotated.limma.2.8.tra <- merge(annotated.limma.2.8, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.2.m$ensembl.transcript = rownames(annotated.limma.2.m)
annotated.limma.2.m.tra <- merge(annotated.limma.2.m, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.2.b$ensembl.transcript = rownames(annotated.limma.2.b)
annotated.limma.2.b.tra <- merge(annotated.limma.2.b, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.4.8$ensembl.transcript = rownames(annotated.limma.4.8)
annotated.limma.4.8.tra <- merge(annotated.limma.4.8, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.4.m$ensembl.transcript = rownames(annotated.limma.4.m)
annotated.limma.4.m.tra <- merge(annotated.limma.4.m, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.4.b$ensembl.transcript = rownames(annotated.limma.4.b)
annotated.limma.4.b.tra <- merge(annotated.limma.4.b, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.8.m$ensembl.transcript = rownames(annotated.limma.8.m)
annotated.limma.8.m.tra <- merge(annotated.limma.8.m, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.8.b$ensembl.transcript = rownames(annotated.limma.8.b)
annotated.limma.8.b.tra <- merge(annotated.limma.8.b, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.m.b$ensembl.transcript = rownames(annotated.limma.m.b)
annotated.limma.m.b.tra <- merge(annotated.limma.m.b, tra.data[,nr], by = 'ensembl.transcript')

```


<!-- Tissue distribution plot -->
```{r Tissue distribution plot, fig.cap= "Tissue distribution plot",fig.height=3.2, fig.width=7, fig.align='center'}
# Generate a list containing all tissue names with number of tissue TRAs, which is detected in our chips
tissue=data.frame(table(tra.data$max.tissue))$Var1

tissue.number = c()

for (i in 1:53) {
  tissue.number[[i]] = nrow(annotated.limma.1.8.tra %>% filter(grepl(tissue[i],annotated.limma.1.8.tra[,13])))
}

tissue.distribution = cbind(as.vector(tissue),tissue.number)
colnames(tissue.distribution)=c("max.tissue","tissue.number")
tissue.distribution = as.data.frame(tissue.distribution)

# Arranging the list according to the descending TRA number
tissue.distribution.arranged = arrange(tissue.distribution, desc(as.numeric(tissue.number)))

# Create a barplot to present the TRA distribution of our chips
jpeg(file="tissue.distribution.arranged.barplot.jpeg")

par(mar=c(5,9,4,1)+.1)
barplot(height = as.numeric(tissue.distribution.arranged$tissue.number[1:10]),names.arg = tissue.distribution.arranged$max.tissue[1:10],cex.names=0.54,col=rainbow(10), main="Frequency of TRA DEGs between 1-cell stage and 8-cell stage", xlab="Frequency",las=2, horiz= TRUE, cex.lab=1, cex.axis= 1)


#setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")

#dev.copy2pdf(file="tissue.distribution.arranged.barplot.pdf" )



dev.off()
```

<!-- #### filter DE genes & volcanoplots -->

In order to visualize our results based on limma and find out interesting TRA genes that play a significant role in embryogenesis, we decided to use Venn Diagram and Volcano plots to process our results.



```{r}
#so far, our limma table only contains significant DEGs. To demonstrate other genes that do not show differential expression we need to create a limma table with all genes. By doing that we found that the ensembl table has some transcripts that repeats itself, so we only select the unique transcripts from the ensembl data
#find out repeating transcripts in ensembl.data
n_occur=data.frame(table(ensembl.data$Transcript.stable.ID))
n_occur=arrange(n_occur,desc(n_occur$Freq))
a=sum(n_occur$Freq>1)
duplicate.ensembl=n_occur$Var1[1:40]
j=which(ensembl.data$Transcript.stable.ID%in%duplicate.ensembl)  
ensembl.duplicate=ensembl.data[j,]
ensembl.duplicate=arrange(ensembl.duplicate,ensembl.duplicate$Transcript.stable.ID)

#delete the repeating transcripts
Nth.delete<-function(dataframe, n)dataframe[-(seq(n,to=nrow(dataframe),by=n)),]
ensembl.duplicate.genes=Nth.delete(ensembl.duplicate, 2)
ensembl.unique=rbind(ensembl.data[-j,],ensembl.duplicate.genes)
```




```{r limma tables with sig & non sig. genes}
#generate a function with the output of limma tables with sig. genes and non sig. genes

generate.limma.table.vollst = function(fit,n){
  fit.1 = contrasts.fit(fit, contrasts = cm[,n])
  fit.1 = eBayes(fit.1)
  limma.table.vollst=topTable(fit.1,number = dim(fit.1)[1])
  return(limma.table.vollst)
}
limma.table.vollst.1.2=generate.limma.table.vollst(fit,1)
limma.table.vollst.1.4=generate.limma.table.vollst(fit,2)
limma.table.vollst.1.8=generate.limma.table.vollst(fit,3)
limma.table.vollst.1.m=generate.limma.table.vollst(fit,4)
limma.table.vollst.1.b=generate.limma.table.vollst(fit,5)
limma.table.vollst.2.4=generate.limma.table.vollst(fit,6)
limma.table.vollst.2.8=generate.limma.table.vollst(fit,7)
limma.table.vollst.2.m=generate.limma.table.vollst(fit,8)
limma.table.vollst.2.b=generate.limma.table.vollst(fit,9)
limma.table.vollst.4.8=generate.limma.table.vollst(fit,10)
limma.table.vollst.4.m=generate.limma.table.vollst(fit,11)
limma.table.vollst.4.b=generate.limma.table.vollst(fit,12)
limma.table.vollst.8.m=generate.limma.table.vollst(fit,13)
limma.table.vollst.8.b=generate.limma.table.vollst(fit,14)
limma.table.vollst.m.b=generate.limma.table.vollst(fit,15)
```


```{r Annotation of limma including all genes}
#anotation of the limma table (complete) with emsembl information
ensembl.limma.annotation <- function(x){
  a = which(ensembl.unique$Transcript.stable.ID %in% rownames(x))
  ensembl.new = ensembl.unique[a,]
  
  b = which(rownames(x)%in% ensembl.new$Transcript.stable.ID)
  x.new = x[b,]
  
  ensembl.new = arrange(ensembl.new,ensembl.new$Transcript.stable.ID)
  x.new = arrange(x.new, rownames(x.new))
  
  annotated.x = cbind(x.new, ensembl.new$Gene.name, ensembl.new$Gene.description,ensembl.new$Chromosome.scaffold.name)
  colnames(annotated.x)[7:9]<-c("gene.name","gene.description","Chromosome.scaffold.name")
  annotated.x = arrange(annotated.x, adj.P.Val)
  return(annotated.x)
}

annotated.limma.table.vollst.1.8 = ensembl.limma.annotation(limma.table.vollst.1.8)
annotated.limma.table.vollst.8.m = ensembl.limma.annotation(limma.table.vollst.8.m) 
annotated.limma.table.vollst.m.b = ensembl.limma.annotation(limma.table.vollst.m.b)
```
### 4.3.1 Venn diagram

To gain an overview of TRA genes that are important for all stages, we use a Venn diagram to plot the intersection of differentially expressed TRA genes from 1-cell stage to 8-cell stage, from 8-cell stage to morula stage as well as from morula stage to blastocyst stage. We extracted genes that are continuously up/down-regulated along the three stage transitions we investigated, since they might be important for the embryogenesis.

```{r Venn diagram up, include=TRUE, fig.cap="Venn diagram upregulated genes",fig.height=3, fig.width=4.5, fig.align='center' }
#creating demo data for Venn diagram
#1.8.tra Upregulated
tra.1.8.up=annotated.limma.1.8.tra[which(annotated.limma.1.8.tra$logFC>0),]$ensembl.transcript
#8.m.tra Upregulated
tra.8.m.up=annotated.limma.8.m.tra[which(annotated.limma.8.m.tra$logFC>0),]$ensembl.transcript
#m.b.tra Upregulated
tra.m.b.up=annotated.limma.m.b.tra[which(annotated.limma.m.b.tra$logFC>0),]$ensembl.transcript

list_venn.up = list(A = tra.1.8.up,B=tra.8.m.up,C=tra.m.b.up)

# Helper function to display Venn diagram.up
display_venn.up <- function(x){
  grid.newpage()
  venn_object <- venn.diagram(x, main=("Upregulated TRAS between different stages"), category.names=c("1-cell and 8-cell stage"," 8-cell and morula stage","morula and blastocyst stage"), fill = c("#009E73", "#E69F00", "#56B4E9"), filename= NULL,cat.cex=0.6)
  grid.draw(venn_object)
}
#,main.fontfamily= "Arial",main.fontface="bold" not working


#venn plot up
display_venn.up(list_venn.up)

#create a function to give the name of the genes in the intersection of three stages
Intersect <- function (x) {  
  # Multiple set version of intersect
  # x is a list
  if (length(x) == 1) {
    unlist(x)
  } else if (length(x) == 2) {
    intersect(x[[1]], x[[2]])
  } else if (length(x) > 2){
    intersect(x[[1]], Intersect(x[-1]))
  }
}

intersect.tra.genes.up=Intersect(list_venn.up)
#9 genes are continuously up-regulated from 1-cell stage to blastocyst stage



Venn.genes.up<-annotated.tra[rownames(annotated.tra) %in% intersect.tra.genes.up,]
```


```{r Venn diagram down, fig.cap="Venn diagram downregulated genes",fig.height=4, fig.width=4, fig.align='center' }
#1.8.tra downregulated
tra.1.8.down=annotated.limma.1.8.tra[which(annotated.limma.1.8.tra$logFC<0),]$ensembl.transcript
#8.m.tra downregulated
tra.8.m.down=annotated.limma.8.m.tra[which(annotated.limma.8.m.tra$logFC<0),]$ensembl.transcript
#m.b.tra downregulated
tra.m.b.down=annotated.limma.m.b.tra[which(annotated.limma.m.b.tra$logFC<0),]$ensembl.transcript

list_venn.down = list(A = tra.1.8.down,B=tra.8.m.down,C=tra.m.b.down)

# Helper function to display Venn diagram.down
display_venn.down <- function(x){
  grid.newpage()
  venn_object.down <- venn.diagram(x, main=("Downregulated TRAS between different stages"), cex.main=2, font.main=1, col.main= "red", category.names=c("1-cell and 8-cell stage"," 8-cell and morula stage","morula and blastocyst stage"),fill = c("#009E73", "#E69F00", "#56B4E9"), filename=NULL)
  grid.draw(venn_object.down)
}
display_venn.down(list_venn.down)

intersect.tra.genes.down=Intersect(list_venn.down)
#5 genes are continuously down-regulated from 1-cell stage to blastocyst stage

Venn.genes.down<-annotated.tra[rownames(annotated.tra) %in% intersect.tra.genes.down,]
```
We found three TRA genes (DAB2, PERP, CHI3L2) that are continuously up-regulated in all three stage transitions and six TRA genes (OTX2, TENT5C, TPRXL, TUBB7P, THBS4, TCL1A) that are continuously down-regulated in all three stage transitions.



### 4.3.2 Volcano plot
After looking that TRA genes that are important for all stages, we then focused on each individual stage transition. In order to display the differential gene expression between each two stages, we created three volcano plots for three stage transitions and colored the significantly differential genes with a p-value cutoff of 0.01 and a fold10change cutoff of 2. To extract the most biologically significant genes, we then identified the top 10 up/down-regulated TRA transcripts that are most significantly expressed and have the largest fold changes. These genes are labeled in the volcano plots.

```{r volcano plots 1.8, include= TRUE, fig.cap= "Volcano plot comparing the 1-cell and 8-cell stage",fig.height=4, fig.width=9, fig.align='center' }
#group the genes into "Up-regulated","Down-regulated" and "Unchanged" for colouring volcano plot. We definded a gene as up/down-regulated wenn the LogFC is higher than 2 or smaller than -2 

#for the stage between 1-cell and 8-cell stage
annotated.limma.table.vollst.1.8 <- annotated.limma.table.vollst.1.8 %>% 
  mutate(
    Expression = case_when(logFC > 2 & adj.P.Val < 0.01 ~ "Up-regulated",
                           logFC < -2 & adj.P.Val < 0.01 ~ "Down-regulated",
                           TRUE ~ "not significantly changed"))


#extract 10 highst up-regulated TRAs and 10 highst down-regulated TRAs arranged by both logFC and adj.P.Value

top <- 10
top_genes_tra.1.8 <- bind_rows(
  annotated.limma.1.8.tra %>% 
    filter(logFC >2) %>% 
    arrange(adj.P.Val,desc(abs(logFC))) %>% 
    head(top),
  annotated.limma.1.8.tra %>% 
    filter(logFC< -2) %>% 
    arrange(adj.P.Val,desc(abs(logFC)))
  %>%head(top)
  )

#top_genes_tra.1.8

#setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
#write.csv(top_genes_tra.1.8, file="top_genes_tra.1.8.csv")

#create a volcano plot
p1 <- ggplot(annotated.limma.table.vollst.1.8, aes(logFC, -log(adj.P.Val,10)))+
  geom_point(aes(color = Expression),size = 2/5)+  
  ggtitle("Volcano plot between 1-cell stage and 8-cell stage")+
  theme(plot.title = element_text(hjust = 0.5,face = "bold",size=14))+
  xlab(expression("log"[10]*"FC")) + 
  ylab(expression("-log"[10]*"(adj.P.Val)"))+
  geom_vline(xintercept=c(-2, 2), col="black",linetype = "dashed") +
  geom_hline(yintercept=-log10(0.01), col="black",linetype = "dashed")

#label the 20 most interesting differencially expressed TRA genes
p1 <-  p1 +
  geom_label_repel(data = top_genes_tra.1.8,
                   mapping = aes(logFC, -log(adj.P.Val,10), label = gene.name),
                   size = 1.5)
p1
```


```{r Volcano plot 8.m, fig.cap= "Volcano plot comparing the 8-cell and morula stage",fig.height=4, fig.width=8, fig.align='center'}
#Repeating the same procedure with the other three stages
#group the genes into "Up-regulated","Down-regulated" and "Unchanged" for colouring volcano plot
annotated.limma.table.vollst.8.m <- annotated.limma.table.vollst.8.m %>% 
  mutate(
    Expression = case_when(logFC >= 2 & adj.P.Val < 0.01 ~ "Up-regulated",
                           logFC <= -2 & adj.P.Val < 0.01 ~ "Down-regulated",
                           TRUE ~ "not significantly changed"))


#extract 10 highst up-regulated TRAs and 10 highst down-regulated TRAs arranged by both logFC and adj.P.Value

top <- 10
top_genes_tra.8.m <- bind_rows(
  annotated.limma.8.m.tra %>% 
    filter(logFC >2) %>% 
    arrange(adj.P.Val,desc(abs(logFC))) %>% 
    head(top),
  annotated.limma.8.m.tra %>% 
    filter(logFC< -2) %>% 
    arrange(adj.P.Val,desc(abs(logFC)))
  %>%head(top)
)
#top_genes_tra.8.m
#setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
#write.csv(top_genes_tra.8.m, file="top_genes_tra.8.m.csv")


#create a volcanoplot
p2 <- ggplot(annotated.limma.table.vollst.8.m, aes(logFC, -log(adj.P.Val,10)))+
  geom_point(aes(color = Expression),size = 2/5)+  
  ggtitle("Volcano plot between 8-cell stage and morula stage")+
  theme(plot.title = element_text(hjust = 0.5,face = "bold",size=10))+
  xlab(expression("log"[10]*"FC")) + 
  ylab(expression("-log"[10]*"(adj.P.Val)"))+
  geom_vline(xintercept=c(-2, 2), col="black",linetype = "dashed") +
  geom_hline(yintercept=-log10(0.01), col="black",linetype = "dashed")

#label the 20 most interesting differentially expressed TRA genes
p2 <-  p2 +
  geom_label_repel(data = top_genes_tra.8.m,
                   mapping = aes(logFC, -log(adj.P.Val,10), label = gene.name),
                   size = 1.5)
p2


```


```{r Volcano plot m.b, fig.cap= " Volcano plot comparing the morula and blastocyst stage",fig.height=4, fig.width=8, fig.align='center'}
#group the genes into "Up-regulated","Down-regulated" and "Unchanged" for colouring volcano plot
annotated.limma.table.vollst.m.b <- annotated.limma.table.vollst.m.b %>% 
  mutate(
    Expression = case_when(logFC >= 2 & adj.P.Val < 0.01 ~ "Up-regulated",
                           logFC <= -2 & adj.P.Val < 0.01 ~ "Down-regulated",
                           TRUE ~ "not significantly changed"))


#extract 10 highst up-regulated TRAs and 10 highst down-regulated TRAs arranged by both logFC and adj.P.Value

top <- 10
top_genes_tra.m.b <- bind_rows(
  annotated.limma.m.b.tra %>% 
    filter(logFC >2) %>% 
    arrange(adj.P.Val,desc(abs(logFC))) %>% 
    head(top),
  annotated.limma.m.b.tra %>% 
    filter(logFC < -2) %>% 
    arrange(adj.P.Val,desc(abs(logFC)))
  %>%head(top)
)
#top_genes_tra.m.b
#setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
#write.csv(top_genes_tra.m.b, file="top_genes_tra.m.b.csv")


#create a volcanoplot
p3 <- ggplot(annotated.limma.table.vollst.m.b, aes(logFC, -log(adj.P.Val,10)))+
  geom_point(aes(color = Expression),size = 2/5)+  
  ggtitle("Volcanoplot between morula stage and blastocyst stage")+
  theme(plot.title = element_text(hjust = 0.5,face = "bold",size=10))+
  xlab(expression("log"[10]*"FC")) + 
  ylab(expression("-log"[10]*"(adj.P.Val)"))+
  geom_vline(xintercept=c(-2, 2), col="black",linetype = "dashed") +
  geom_hline(yintercept=-log10(0.01), col="black",linetype = "dashed")

#label the 20 most interesting differentially expressed TRA genes
p3 <-  p3 +
  geom_label_repel(data = top_genes_tra.m.b,
                   mapping = aes(logFC, -log(adj.P.Val,10), label = gene.name),
                   size = 1.5)
p3
```


Interestingly, TRA genes that are related to gametes and gametogenesis (xxx and xxx) are found down-regulated between 1-cell stage and 8-cell stage. Between 8-m stage and m-b stage, genes that activate oxidative phosphorylation and help against oxidative stress, like xxx and xxx, are found up-regulated. At the same time, xxx and xxx are found highly up-regulated, which are related to preimplantation.


<!-- #Compare our top TRA genes that we just extracted with those TRA genes, which have the highest loading in PCs and explain for the variance between different stages. -->


Finding TRA genes which explain the variance by looking at TRA genes with the highest loading in PCs

<!-- # Use PCA to plot 3 replicates of 1-cell stage and 8-cell stage -->

```{r Extracting TRA genes explaining variance (1-8)}
#Use PCA to plot 3 replicates of 1-cell stage and 8-cell stage
a=c(1,2,3,10,11,12)
pca.1.8=prcomp(t(human.vsnrma.df2[,a]))

# Plot cumulative proportion of variance explained
variance = (pca.1.8$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
#two PCs are enough to show 90% variance, we will just focus on PC1

#Looking for TRA genes which explain the variance of PC1 between 1-cell stage and 8-cell stage
pc1=pca.1.8$rotation[,1]
top_genes_pca1.1.8 = names(sort(abs(pc1), decreasing = T))[1:70]

#extract 20 top TRA genes that explains the variance between 1-cell stage and 8-cell stage
k = which(rownames(annotated.tra)%in% top_genes_pca1.1.8)
top_tra_pca1.1.8=annotated.tra[k,]
dim(top_tra_pca1.1.8)
#20 25

length(unique(top_genes_tra.1.8$gene.name))
# 15
intersect(top_genes_tra.1.8$gene.name,top_tra_pca1.1.8$Gene.name)
# 7 of our top TRA genes (1.8) are present in the variance explaining genes 
#"KDM5B"   "IMPDH1"  "KHDC1L"  "IER3"    "ZNF280A" "C4orf47" "ZNF571"
# 7/15 = 46.67%
```

<!-- # Use PCA to plot 3 replicates of 8-cell stage and morula stage -->

```{r Extracting TRA genes explaining variance (8-m)}
#Use PCA to plot 3 replicates of 8-cell stage and morula stage
pca.8.m=prcomp(t(human.vsnrma.df2[,10:15]))

# Plot cumulative proportion of variance explained
plot(pca.8.m$sdev)
variance = (pca.8.m$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)

#three PCs are enough to show 90% variance, we will just focus on PC1

#Looking for TRA genes which explain the variance of PC1 between 8-cell stage and morula stage
pc1=pca.8.m$rotation[,1]
top_genes_pca1.8.m = names(sort(abs(pc1), decreasing = T)[1:35])

#extract 20 top TRA genes that explains the variance between 8-cell stage and morula stage
k = which(rownames(annotated.tra)%in% top_genes_pca1.8.m)
top_tra_pca1.8.m=annotated.tra[k,]
dim(top_tra_pca1.8.m)
#20 25

length(unique(top_genes_tra.8.m$gene.name))
# 9
intersect(top_genes_tra.8.m$gene.name,top_tra_pca1.8.m$Gene.name)
# 3 of our top TRA genes (8.m) are present in the variance explaining genes 
# "SLC2A3" "PCP4L1" "RIIAD1"
# 3/9 = 33.33%
```


<!-- *Use PCA to plot 3 replicates of morula stage and blastocyst stage* -->

```{r}
#Use PCA to plot 3 replicates of morula stage and blastocyst stage
pca.m.b=prcomp(t(human.vsnrma.df2[,13:18]))

# Plot cumulative proportion of variance explained
plot(pca.m.b$sdev)
variance = (pca.m.b$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
#four PCs are enough to show 90% variance, we will just focus on PC1

#Looking for TRA genes which explain the variance of PC1 between morula stage and blastocyst stage
pc1=pca.m.b$rotation[,1]
top_genes_pca1.m.b = names(sort(abs(pc1), decreasing = T)[1:55])

#extract 20 top TRA genes that explains the variance between 8-cell stage and morula stage
k = which(rownames(annotated.tra)%in% top_genes_pca1.m.b)
top_tra_pca1.m.b=annotated.tra[k,]
dim(top_tra_pca1.m.b)
#20 25

length(unique(top_genes_tra.m.b$gene.name))
# 11
intersect(top_genes_tra.m.b$gene.name,top_tra_pca1.m.b$Gene.name)
# 5 of our top TRA genes (m.b) are present in the variance explaining genes 
# "ABCG2"    "NQO1"     "ATP6V0A4" "ZNF280A"  "ELL3"    
# 5/11 = 45.45%
```


### 4.3.3 Gene set enrichment analysis
After extracting the top up/down-regulated DEGs and understanding their biological relevance in embryo as well as their presence in possible pathways, we decided to perform GSEA to find out if our DEGs were generally enriched in certain gene sets of biological pathways. (Therefore, we compared the DEGs with gene sets of the Gene Ontology (GO) database using the limma function goana(). Then, we generated topTables with the results of GSEA that contain the significantly enriched or depleted biological pathways. In addition, these tables provide information about the pathway ID and name, the number of genes in the set, the number of up- and downregulated genes, and the p-value for enrichment of up- and downregulated genes.)
 
(eher Methode?)

```{r GSEA for stage m-b,include=TRUE, fig.cap="Gene set enrichment analysis comparing morula and blastocyst stage", fig.height=4, fig.width=4}

organism = "org.Hs.eg.db"

#create the input for the function gseGO() for stage m-b
# we want the log10 fold change 
GO.m.b.genes<- annotated.limma.m.b$logFC

# name the vector with the Transkript ID
names(GO.m.b.genes) <- rownames(annotated.limma.m.b)


# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(GO.m.b.genes, decreasing = TRUE)

# conduct the GSEA with GO data
gse.m.b.DEG <- gseGO(geneList=gene_list, 
                 ont ="ALL", 
                 keyType = "ENSEMBLTRANS", 
                 minGSSize = 3, 
                 maxGSSize = 800, 
                 pvalueCutoff = 0.05, 
                 verbose = TRUE, 
                 OrgDb = organism, 
                 pAdjustMethod = "BH")

#visualize the GSEA output
require(DOSE)
dotplot(gse.m.b.DEG, showCategory=10, split=".sign",font.size=3) + facet_grid(.~.sign)
#gse.m.b.DEG@result
```






## 4.4 Chemokines
According to our hypothesis, we also examined the expression of chemokines. The result of the filtering of our annotated TRA matrix showed that there are 62 proteins with a chemokine motif. Due to this, we looked at whether some of the filtered proteins are differentially expressed genes (DEGs) in our four clusters. The result of this is that there are no proteins with a chemokine motif that are TRAs and DEGs. 

Thus, we looked after proteins with a chemokine motif in all of our DEGs including non-TRA genes. We got the following results:

+ Between the stages 1-cell to 8-cell CXCL14 was upregulated with a log fold change (logFC) of approximately 1 and the chemokine receptor CCR3 was downregulated (logFC= -1.12). 
+ Between the stages 8-cell to morula the expression of  CCL15 was upregulated with a logFC of approximately 2.45 whereas in morula to blastocyst the expression was downregulated with a logFC of -2.66

```{r chemokine analysis}
#extract chemokine in our annotated tra table
chemokine.tra.list=annotated.tra[which(grepl("motif chemokine",annotated.tra[,24])),]
dim(chemokine.tra.list)
n_occur=data.frame(table(chemokine.tra.list$Chromosome.scaffold.name))
print(n_occur)
#62 25
chemokine.list=ensembl.data[which(grepl("motif chemokine",ensembl.data[,8])),]
dim(chemokine.list)
#280 9

#create a function to extract differentially expressed TRA chemokines from limma table
extract.chemokine.tra <- function(x){
x.chemokine.tra=x[which(rownames(x) %in% chemokine.tra.list$Transcript.stable.ID),]
  return(x.chemokine.tra)
}

annotated.limma.1.8.chemokine.tra=extract.chemokine.tra(annotated.limma.1.8)
dim(annotated.limma.1.8.chemokine.tra)
#0 9
annotated.limma.8.m.chemokine.tra=extract.chemokine.tra(annotated.limma.8.m)
dim(annotated.limma.8.m.chemokine.tra)
#0 9
annotated.limma.m.b.chemokine.tra=extract.chemokine.tra(annotated.limma.m.b)
dim(annotated.limma.m.b.chemokine.tra)
#0 9

#no TRA chemokines found 

#create a function to extract differentially expressed chemokines (non TRA) from limma table
extract.chemokine <- function(x){
  x.chemokine=x[which(rownames(x) %in% chemokine.list$Transcript.stable.ID),]
  return(x.chemokine)
}

annotated.limma.1.8.chemokine=extract.chemokine(annotated.limma.1.8)
dim(annotated.limma.1.8.chemokine)
#6 11

unique(annotated.limma.1.8.chemokine$gene.name)
#"CCR3" Down  "CXCL14" Up

annotated.limma.8.m.chemokine=extract.chemokine(annotated.limma.8.m)
dim(annotated.limma.8.m.chemokine)
#4 11

unique(annotated.limma.8.m.chemokine$gene.name)
#"CCL15" Up


annotated.limma.m.b.chemokine=extract.chemokine(annotated.limma.m.b)
dim(annotated.limma.m.b.chemokine)
#4 11


unique(annotated.limma.m.b.chemokine$gene.name)
#"CCL15" Down

```

```{r chemokine heatmap,include=TRUE,fig.cap="Chemokine heatmap",fig.height=3, fig.width=5}
#extract the chemokine transkript
chemokine.transkript=unique(c(rownames(annotated.limma.1.8.chemokine),rownames(annotated.limma.8.m.chemokine),rownames(annotated.limma.m.b.chemokine)))

#extract the chemokine information from ensembl table
chemokine.deg=ensembl.data[which(ensembl.data$Transcript.stable.ID %in% chemokine.transkript),]


#create the chemokine expression matrix
chemokine.expression=human.vsnrma.df2[which(rownames(human.vsnrma.df2) %in% chemokine.transkript),]

#create the heatmap
annotation_row = data.frame(
  gene.name = chemokine.deg$Gene.name
)

rownames(annotation_row) = chemokine.deg$Transcript.stable.ID

heatmap=pheatmap(as.matrix(chemokine.expression[,1:18]),main = "Expression of chemokines",cluster_cols=FALSE,cluster_rows=TRUE, show_rownames = TRUE, legend=TRUE, fontsize_row=0.5, annotation_row = annotation_row)

```

# 5. Discussion

## 5.1	General embryo development 

## 5.2	Maternal to zygotic transition  
In the early stage of embryonic development, the embryo is transcriptionally silent, which means, that the genome activation of the zygote occurs a bit later. Just after the so-called development stage of maternal to zygotic transition (MZT), the newly arranged genom from the sperm and maternal egg is used. Therefore, degradation of maternal mRNA needs to take place (Schier, 2007).

From the results, it is clear that in our data the MZT takes place in between stages 1 cell to 8 cell. This can be seen in the differential expression of the proteins like PRPF4B and EIF5. After the transcription the non-coding sequences of the RNA need to be removed by the process called RNA-Splicing. PRPF4B has a regulative role in the pre-mRNA splicing (Corkery et al., 2015).    
EIF5 plays an important role in the formation of a functional 80 S initiation complex (Si et al., 1996).  Therefore an up regulation of both proteins indicates translation events which also imply that the activation of the zygotic genome took place.  
 
 


## 5.3	Switch of energy metabolism

During our limma analysis of the differentially expressed genes we also found genes indicating a change of energy source in the mammalian embryo development. In the late morula stage glucose becomes the predominant exogenous energy substrate while before the energy was mostly derived from aerobic substrates like pyruvate, lactate, and glutamine (Leese et al.,1993). At the same time the oxygen consumption increases. This is a result of ATP synthesis in the trophectoderm which is driven by oxidative phosphorylation. Unfortunately, the production of reactive oxygen species (ROS) thereby also rises (Kaneko et al., 2013). One indicator for this event is that in our data the SLC2A3 gene was significantly upregulated in the morula stage. SLC2A3 encodes for a glucose transporter (GLUT3) which is responsible for diffusion of glucose across plasma membranes (Ziegler et al., 2020). Furthermore, we found genes encoding for proteins which prevent excess accumulation of ROS during oxidative phosphorylation (TEAD4)(Kaneko et al., 2013) and proteins which defense the cell against oxidative stress (NQO1)(Ross et al.,2021).This energy homeostasis is thought to be caused by increased energy demands due to two processes: the pumping of sodium ions by the Na+/K+ ATPase in the trophectoderm and the increased protein synthesis because the embryo starts its net growth (Leese et al., 2008).

## 5.4	Implantation 
Even though our data only includes preimplantation stages, we still can observe the first differentiative event in embryogenesis- the formation of the blastocyst. The outer cells start to form tight junctions and build an epithelial layer which is called the trophectoderm. This layer is later involved in in the implantation process meaning the blastocyst will settle in the uterine endometrium (Leese et al.,1993). To ensure a successful implantation several arrangements must be made in advance. One of them is the expression of STMN1 which promotes the migration and invasion of the trophoblast by regulating the expression of N-cadherin and E-cadherin. N-cadherin promotes invasion whereas E-cadherin suppresses the invasion (Tian et al.,2015). In our data one can observe a significant increase in STMN1 expression during the morula stage which indicates that the preparation for implantation starts long before the actual implantation event. 

As described in the introduction, chemokines expression are durin important during the embryonic development. For successful implantation, new blood vessels need to be formed in the endometrium (Torry et al., 2007). CCL15 showed to induce the process of angiogenesis (Hwang et al., 2004). Our analysis showed that this chemokine is first upregulated between the stages 8-cell and morula. This also demonstrates the process of implantation preparation. However, we could not explain the downregulation between the stages morula and blastocyst. Our findings on the upregulation of CXCL14 hint that it could also play a significant role in embryogenesis. This chemokine is related to CXCL12 whose expression showed to encourage angiogenesis and a positive pregnancy course (Koo et al., 2021). In line with previous studies, where the involvement of CXCL14 in the same embryonic pathways as CXCL12 (Nassari et al., 2017), the expressions seem logical. All in all, it can be said, that our analysis points up that chemokines play a role in early embryonc development. But ensuing steps for further analysis and research should be taken, to study the mechanism by which those signal proteins are involved in the embryogenesis.

\newpage
# 6. Supplementary material
## 6.1 Top 10 upregualted TRA genes in the compared stages
```{r toptable 1.8, include= TRUE}
#Top table
f<-c(1,2,8)

knitr::kable(top_genes_tra.1.8[1:10,f],format = "simple",caption ="Upregulated genes comparing 1-cell and 8-cell stage")

```

```{r toptable 8.m, include= TRUE}
#Top table

knitr::kable(top_genes_tra.8.m[1:10,f],format = "simple",caption ="Upregulated genes comparing 8-cell and morula stage")

```


```{r toptable m.b, include= TRUE}
#Top table

knitr::kable(top_genes_tra.m.b[1:10,f],format = "simple",caption ="Upregulated genes comparing morula and blastocyst stage")

```

### 6.1.1 Expression plots of certain top 10 genes

```{r Expression plot STMN1, include=TRUE, fig.cap="Expression plot of Stathmin 1", fig.height=3,fig.width=6}
names1 = c("1 ", 
          "2 ",
          "4 ",
          "8 ",
          "m ",
          "b")


annotated.tra.mean= cbind(rownames(annotated.tra),rowMeans(annotated.tra[,1:3]),rowMeans(annotated.tra[,4:6]),rowMeans(annotated.tra[,7:9]),rowMeans(annotated.tra[,10:12]),rowMeans(annotated.tra[,13:15]),rowMeans(annotated.tra[,16:18]),annotated.tra[,19:25])

colnames(annotated.tra.mean)[2:7]=names1

o = c(which(annotated.tra.mean[,12] %in% "STMN1"))

annotated.tra.mean.1=pivot_longer(annotated.tra.mean[o,], cols=2:7, names_to = "Stages" ,values_to="expression")

colnames(annotated.tra.mean.1)[1]= "transcript"

ggplot(annotated.tra.mean.1, aes(x = Stages, y =expression,group=as.character(transcript))) +
  geom_line(aes(color=transcript)) + scale_x_discrete(limits=names1) +
  labs(y=("Expression")) +
    ggtitle("Expression STMN1")
```

```{r Expression plot SLC2A3,include= TRUE, fig.cap="Expression plot of Solute Carrier Family 2 Member 3", fig.height=3,fig.width=6}
names1 = c("1 ", 
          "2 ",
          "4 ",
          "8 ",
          "m ",
          "b")


annotated.tra.mean= cbind(rownames(annotated.tra),rowMeans(annotated.tra[,1:3]),rowMeans(annotated.tra[,4:6]),rowMeans(annotated.tra[,7:9]),rowMeans(annotated.tra[,10:12]),rowMeans(annotated.tra[,13:15]),rowMeans(annotated.tra[,16:18]),annotated.tra[,19:25])

colnames(annotated.tra.mean)[2:7]=names1

o = c(which(annotated.tra.mean[,12] %in% "SLC2A3"))

annotated.tra.mean.1=pivot_longer(annotated.tra.mean[o,], cols=2:7, names_to = "Stages" ,values_to="expression")

colnames(annotated.tra.mean.1)[1]= "transcript"

ggplot(annotated.tra.mean.1, aes(x = Stages, y =expression,group=as.character(transcript))) +
  geom_line(aes(color=transcript)) + scale_x_discrete(limits=names1) +
  labs(y=("Expression")) +
    ggtitle("Expression SLC2A3")
```

```{r Expression plot TEAD4, include=TRUE, fig.cap="Expression plot of TEA Domain Transcription Factor 4", fig.height=3,fig.width=6}
names1 =c("1 ", 
          "2 ",
          "4 ",
          "8 ",
          "m ",
          "b")

annotated.tra.mean= cbind(rownames(annotated.tra),rowMeans(annotated.tra[,1:3]),rowMeans(annotated.tra[,4:6]),rowMeans(annotated.tra[,7:9]),rowMeans(annotated.tra[,10:12]),rowMeans(annotated.tra[,13:15]),rowMeans(annotated.tra[,16:18]),annotated.tra[,19:25])

colnames(annotated.tra.mean)[2:7]=names1

o = c(which(annotated.tra.mean[,12] %in% "TEAD4"))

annotated.tra.mean.1=pivot_longer(annotated.tra.mean[o,], cols=2:7, names_to = "Stages" ,values_to="expression")

colnames(annotated.tra.mean.1)[1]= "transcript"

ggplot(annotated.tra.mean.1, aes(x = Stages, y =expression,group=as.character(transcript))) +
  geom_line(aes(color=transcript)) + scale_x_discrete(limits=names1) +
  labs(y=("Expression")) +
    ggtitle("Expression TEAD4")
```

```{r Expression plot NQO1, include=TRUE, fig.cap="Expression plot of NAD(P)H Quinone Dehydrogenase 1", fig.height=3,fig.width=6}
names1 = c("1 ", 
          "2 ",
          "4 ",
          "8 ",
          "m ",
          "b")


annotated.tra.mean= cbind(rownames(annotated.tra),rowMeans(annotated.tra[,1:3]),rowMeans(annotated.tra[,4:6]),rowMeans(annotated.tra[,7:9]),rowMeans(annotated.tra[,10:12]),rowMeans(annotated.tra[,13:15]),rowMeans(annotated.tra[,16:18]),annotated.tra[,19:25])

colnames(annotated.tra.mean)[2:7]=names1

o = c(which(annotated.tra.mean[,12] %in% "NQO1"))

annotated.tra.mean.1=pivot_longer(annotated.tra.mean[o,], cols=2:7, names_to = "Stages" ,values_to="expression")

colnames(annotated.tra.mean.1)[1]= "transcript"

ggplot(annotated.tra.mean.1, aes(x = Stages, y =expression,group=as.character(transcript))) +
  geom_line(aes(color=transcript)) + scale_x_discrete(limits=names1) +
  labs(y=("Expression")) +
    ggtitle("Expression NQO1")
```
## 6.2 Gene set enrichment analysis


```{r GSEA for stage 1-8, include=TRUE, fig.cap="Gene set enrichment analysis comparing 1-cell and 8-cell stage"}
#create the input for the function gseGO() for stage 1-8
# we want the log10 fold change 
GO.1.8.genes<- annotated.limma.1.8$logFC

# name the vector with Transkript ID
names(GO.1.8.genes) <- rownames(annotated.limma.1.8)


# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(GO.1.8.genes, decreasing = TRUE)

# conduct the GSEA with GO data
organism = "org.Hs.eg.db"
gse.1.8.DEG <- gseGO(geneList=gene_list, 
                 ont ="ALL", 
                 keyType = "ENSEMBLTRANS", 
                 minGSSize = 3, 
                 maxGSSize = 800, 
                 pvalueCutoff = 0.05,
                 verbose = TRUE, 
                 OrgDb = organism, 
                 pAdjustMethod = "BH")

#visualize the GSEA output
require(DOSE)
dotplot(gse.1.8.DEG, showCategory=10, split=".sign",font.size=5) + facet_grid(.~.sign)
#gse.1.8.DEG@result
```

```{r GSEA for stage 8-m, include=TRUE, fig.cap="Gene set enrichment analysis comparing 8-cell and morula stage"}
#create the input for the function gseGO() for stage 8-m
# we want the log10 fold change 
GO.8.m.genes<- annotated.limma.8.m$logFC

# name the vector with Transkript ID
names(GO.8.m.genes) <- rownames(annotated.limma.8.m)


# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(GO.8.m.genes, decreasing = TRUE)

# conduct the GSEA with GO data
gse.8.m.DEG <- gseGO(geneList=gene_list, 
                 ont ="ALL", 
                 keyType = "ENSEMBLTRANS", 
                 minGSSize = 3, 
                 maxGSSize = 800, 
                 pvalueCutoff = 0.05, 
                 eps=0,
                 verbose = TRUE, 
                 OrgDb = organism, 
                 pAdjustMethod = "BH")

#visualize the GSEA output
require(DOSE)
dotplot(gse.8.m.DEG, showCategory=10, split=".sign",font.size=4) + facet_grid(.~.sign)
#gse.8.m.DEG@result
```


\newpage
# 7. References


Leese, H.J., Conaghan, J., Martin, K.L., and Hardy, K. (1993). Early human embryo metabolism. Bioessays 15, 259-264. 10.1002/bies.950150406.

Corkery, D.P., Holly, A.C., Lahsaee, S., and Dellaire, G. (2015). Connecting the speckles: Splicing kinases and their role in tumorigenesis and treatment response. Nucleus 6, 279-288. 10.1080/19491034.2015.1062194.

Kaneko, K.J., and DePamphilis, M.L. (2013). TEAD4 establishes the energy homeostasis essential for blastocoel formation. Development 140, 3680-3690. 10.1242/dev.093799.

Bortz, J., Schuster, C. (2010). Statistik für Human- und Sozialwissenschaftler, 7 Edition (Springer Berlin, Heidelberg). 10.1007/978-3-642-12770-0.

Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43, e47-e47. 10.1093/nar/gkv007.

Subramanian, A., Tamayo, P., Mootha, V.K., Mukherjee, S., Ebert, B.L., Gillette, M.A., Paulovich, A., Pomeroy, S.L., Golub, T.R., Lander, E.S., and Mesirov, J.P. (2005). Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles. Proceedings of the National Academy of Sciences 102, 15545-15550. doi:10.1073/pnas.0506580102.

Wang, W., Dong, J., Chen, B., Du, J., Kuang, Y., Sun, X., Fu, J., Li, B., Mu, J., Zhang, Z., et al. (2020). Homozygous mutations in REC114 cause female infertility characterised by multiple pronuclei formation and early embryonic arrest. J Med Genet 57, 187-194. 10.1136/jmedgenet-2019-106379.

Schier, A.F. (2007). The maternal-zygotic transition: death and birth of RNAs. Science 316, 406-407. 10.1126/science.1140693.

Leese, H.J., Baumann, C.G., Brison, D.R., McEvoy, T.G., and Sturmey, R.G. (2008). Metabolism of the viable mammalian embryo: quietness revisited. Mol Hum Reprod 14, 667-672. 10.1093/molehr/gan065.

Ziegler, G.C., Almos, P., McNeill, R.V., Jansch, C., and Lesch, K.P. (2020). Cellular effects and clinical implications of SLC2A3 copy number variation. J Cell Physiol 235, 9021-9036. 10.1002/jcp.29753.

Ross, D., and Siegel, D. (2021). The diverse functionality of NQO1 and its roles in redox control. Redox Biol 41, 101950. 10.1016/j.redox.2021.101950.

Tian, F.J., Qin, C.M., Li, X.C., Wu, F., Liu, X.R., Xu, W.M., and Lin, Y. (2015). Decreased stathmin-1 expression inhibits trophoblast proliferation and invasion and is associated with recurrent miscarriage. Am J Pathol 185, 2709-2721. 10.1016/j.ajpath.2015.06.010.

Huber, W., von Heydebreck, A., Sültmann, H., Poustka, A., and Vingron, M. (2002). Variance stabilization applied to microarray data calibration and to the quantification of differential expression. Bioinformatics 18 Suppl 1, S96-104. 10.1093/bioinformatics/18.suppl_1.s96.

Xie, D., Chen, C.C., Ptaszek, L.M., Xiao, S., Cao, X., Fang, F., Ng, H.H., Lewin, H.A., Cowan, C., and Zhong, S. (2010). Rewirable gene regulatory networks in the preimplantation embryonic development of three mammalian species. Genome Res 20, 804-815. 10.1101/gr.100594.109.

Gentleman, R., Carey, V. J., Huber, W., Irizarry, R. A., Dudoit, S. . (2005). Bioinformatics and Computational Biology Solutions Using R and Bioconductor (Springer New York, NY). 10.1007/0-387-29362-0.


Leese, H.J., Conaghan, J., Martin, K.L., and Hardy, K. (1993). Early human embryo metabolism. Bioessays 15, 259-264. 10.1002/bies.950150406.

Corkery, D.P., Holly, A.C., Lahsaee, S., and Dellaire, G. (2015). Connecting the speckles: Splicing kinases and their role in tumorigenesis and treatment response. Nucleus 6, 279-288. 10.1080/19491034.2015.1062194.

Kaneko, K.J., and DePamphilis, M.L. (2013). TEAD4 establishes the energy homeostasis essential for blastocoel formation. Development 140, 3680-3690. 10.1242/dev.093799.

Bortz, J., Schuster, C. (2010). Statistik für Human- und Sozialwissenschaftler, 7 Edition (Springer Berlin, Heidelberg). 10.1007/978-3-642-12770-0.

Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43, e47-e47. 10.1093/nar/gkv007.

Subramanian, A., Tamayo, P., Mootha, V.K., Mukherjee, S., Ebert, B.L., Gillette, M.A., Paulovich, A., Pomeroy, S.L., Golub, T.R., Lander, E.S., and Mesirov, J.P. (2005). Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles. Proceedings of the National Academy of Sciences 102, 15545-15550. doi:10.1073/pnas.0506580102.

Wang, W., Dong, J., Chen, B., Du, J., Kuang, Y., Sun, X., Fu, J., Li, B., Mu, J., Zhang, Z., et al. (2020). Homozygous mutations in REC114 cause female infertility characterised by multiple pronuclei formation and early embryonic arrest. J Med Genet 57, 187-194. 10.1136/jmedgenet-2019-106379.

Schier, A.F. (2007). The maternal-zygotic transition: death and birth of RNAs. Science 316, 406-407. 10.1126/science.1140693.

Leese, H.J., Baumann, C.G., Brison, D.R., McEvoy, T.G., and Sturmey, R.G. (2008). Metabolism of the viable mammalian embryo: quietness revisited. Mol Hum Reprod 14, 667-672. 10.1093/molehr/gan065.

Ziegler, G.C., Almos, P., McNeill, R.V., Jansch, C., and Lesch, K.P. (2020). Cellular effects and clinical implications of SLC2A3 copy number variation. J Cell Physiol 235, 9021-9036. 10.1002/jcp.29753.

Ross, D., and Siegel, D. (2021). The diverse functionality of NQO1 and its roles in redox control. Redox Biol 41, 101950. 10.1016/j.redox.2021.101950.

Tian, F.J., Qin, C.M., Li, X.C., Wu, F., Liu, X.R., Xu, W.M., and Lin, Y. (2015). Decreased stathmin-1 expression inhibits trophoblast proliferation and invasion and is associated with recurrent miscarriage. Am J Pathol 185, 2709-2721. 10.1016/j.ajpath.2015.06.010.

Huber, W., von Heydebreck, A., Sültmann, H., Poustka, A., and Vingron, M. (2002). Variance stabilization applied to microarray data calibration and to the quantification of differential expression. Bioinformatics 18 Suppl 1, S96-104. 10.1093/bioinformatics/18.suppl_1.s96.

Xie, D., Chen, C.C., Ptaszek, L.M., Xiao, S., Cao, X., Fang, F., Ng, H.H., Lewin, H.A., Cowan, C., and Zhong, S. (2010). Rewirable gene regulatory networks in the preimplantation embryonic development of three mammalian species. Genome Res 20, 804-815. 10.1101/gr.100594.109.

Gentleman, R., Carey, V. J., Huber, W., Irizarry, R. A., Dudoit, S. . (2005). Bioinformatics and Computational Biology Solutions Using R and Bioconductor (Springer New York, NY). 10.1007/0-387-29362-0.

Hannan, N.J., and Salamonsen, L.A. (2007). Role of chemokines in the endometrium and in embryo implantation. Current Opinion in Obstetrics and Gynecology 19, 266-272. 10.1097/GCO.0b013e328133885f.


Raman, D., Sobolik-Delmaire, T., and Richmond, A. (2011). Chemokines in health and disease. Exp Cell Res 317, 575-589. 10.1016/j.yexcr.2011.01.005.

Torry, D.S., Leavenworth, J., Chang, M., Maheshwari, V., Groesch, K., Ball, E.R., and Torry, R.J. (2007). Angiogenesis in implantation. J Assist Reprod Genet 24, 303-315. 10.1007/s10815-007-9152-7.

Hwang, J., Kim, C.W., Son, K.N., Han, K.Y., Lee, K.H., Kleinman, H.K., Ko, J., Na, D.S., Kwon, B.S., Gho, Y.S., and Kim, J. (2004). Angiogenic activity of human CC chemokine CCL15 in vitro and in vivo. FEBS Lett 570, 47-51. 10.1016/j.febslet.2004.06.023.

Koo, H.S., Yoon, M.-J., Hong, S.-H., Ahn, J., Cha, H., Lee, D., Ko, J.-E., Kwon, H., Choi, D.H., Lee, K.-A., et al. (2021). CXCL12 enhances pregnancy outcome via improvement of endometrial receptivity in mice. Scientific Reports 11, 7397. 10.1038/s41598-021-86956-y.

Nassari, S., Blavet, C., Bonnin, M.A., Stricker, S., Duprez, D., and Fournier-Thibault, C. (2017). The chemokines CXCL12 and CXCL14 differentially regulate connective tissue markers during limb development. Sci Rep 7, 17279. 10.1038/s41598-017-17490-z.


