---
title: "Final Report "
date: "18.07.2022"
output:
  pdf_document: default
  html_document:
    df_print: paged
header-includes:
   - \usepackage{float}
   - \usepackage{caption}
   - \captionsetup[figure]{font=footnotesize}
---
\thispagestyle{empty}
\hrule
\vspace{0.3cm}
\begin{center}
	\vspace{1cm}
  \large
	\begin{tabular}[c]{l}
	 \\

Clara Certa, Yaxin Chen, Linda Kaupp, Alewtina Towara \\
\\
	 \\
	Supervisor:
	Dr. Maria Dinkelacker, Dr. Carl Herrmann \\
	 \\
	Tutor:
	Ian Dirk Fichter \\
	 \\
	 \\
	Data Analysis for students of Molecular Biotechnology \\
	\\
	Heidelberg University 
	\vspace{0.3cm} \\
	\end{tabular}
	\end{center}

---

\pagenumbering{gobble}
\pagebreak
\tableofcontents
\pagebreak
\pagenumbering{arabic}


```{r setup, include=FALSE}
knitr::opts_chunk$set(include=FALSE, echo = FALSE, fig.align = 'centre', message=FALSE, cache = TRUE)
knitr::opts_knit$set(root.dir = "~//Documents//GitHub//2022-topic-04-team-03")
```

```{r}
projectPath <- dirname(rstudioapi::getSourceEditorContext()$path)
```

# Table of contents
1 [Abstract](#abstarct) 

\newpage

# 1.Abstract
Research, especially in the field of early embryogenesis is important in many ways. It is not only required for better understanding of genetic defects of the embryos in an early stage it is also important for improvement of assisted human conception. In this report we are investigating the preimplantation stages of embryo development up to the formation of the blastocyst. As the implantation only occurs after the stages displayed by our dataset, we expected to not observe any tissue specific development and therefore no differentially expressed tissue-restricted antigens (TRAs) during this period. In contrast we expected expression of chemokines involved in migration and implantation in the uterine endometrium.
\newpage

# 2.Introduction
After the fertilization the human embryo begins to cleave every 24 hours. The individual cells can be detected until the 16-cell stage. Till then every single cell is able to form a whole fetus as well as the fetal placenta. After the 16-cell stage an event called compaction occurs. The individual cells, or blastomeres, become flat and thereby maximize the cell-cell contacts to one another. The result of compaction is the morula stage. The next step is the formation of the blastocyst which is the first differentiative step in early embryo development. The outer cells form an epithelial layer around the embryo called the trophectoderm. The trophectoderm is later important for implantation in the uterine endometrium. As soon as the epithelial layer is built those cells start the ingestion of fluid which leads to the formation of the blastocoel cavity. The blastocoel is additionally containing a small number of undefined cells which will later form the actual embryo. Those cells are referred to as the inner cell mass (Leese et al.,1993).
Interestingly the activation of the embryo genome is thought to only start after the 4-cell stage. Prior the development is driven by maternal stores of RNA which were passed during oogenesis. 
The metabolic activity itself only starts rising at the late morula stage. The two reasons for that are the increased energy demand caused by protein biosynthesis as the number of dividing cells goes up and the ion pumping into the blastocoel cavity. 
During this phase the energy metabolism of the embryo changes from using aerobic substrates such as pyruvate, lactate, and glutamine to using glucose (Leese et al.,1993). 
This switch results in a 2.7-fold rise of oxygen consumption in the trophectoderm where the oxidative phosphorylation takes place. As a consequence, the production of harmful reactive oxygen species (ROS) increases. The embryo therefore has to invent a system to prevent high amounts of ROS and at the same time prepare repair mechanisms for fixing the damage (Kaneko et al.,2013).

\newpage

# 3.Materials and methods
##3.1 Datasets
In order to study the cellular and tissue development during human early embryogenesis we analyzed a microarray dataset published by Xie et al.in 2010. We examined the amount of differentially expressed tissue-restricted antigens (TRAs) in our dataset by comparing it with the TRA dataset provided by Dr. Dinkelacker in 2017.

###3.1.1 Microarray dataset
The microarray dataset by Xie et all. contained three datasets one of bovine, one of mouse and one of human. We decided to focus our project on the human dataset. The data was generated by RNA extraction, amplification, and hybridisation onto Affymetrix microarrays. For early human embryogenesis they prepared three replicates at one cell stage, two cell stage, four cell stage, eight cell stage, morula, and blastocyst each. Each replicate included 95659 transcripts.

###3.1.2 TRA dataset
As a reference TRA dataset we chose the GTEX RNAseq human dataset provided by Dr. Dinkelacker in 2017. It consisted of 60131 transcripts.





```{r load packages, eval=TRUE, include=FALSE, comment=FALSE}
#load library
library(affy)
library(vsn)
library(AnnotationDbi)
library(hgu133plus2hsenstcdf)
library(hgu133plus2hsenstprobe)
library(limma)
library(pheatmap)
library(dplyr)
library(tidyverse)
library(hexbin)
library(cluster)
library(factoextra)
library(EnhancedVolcano)
library(org.Hs.eg.db)
library(RColorBrewer)
library(VennDiagram)
library(GO.db)
library(tinytex)
library(clusterProfiler)
library(enrichplot)
library(ggplot2)
```



## 3.2 Quality Control and Normalization of the Data

Before analyzing the microarrays, it is important to determine the quality of the used raw data. This is done by the procedure called Quality Control (QC) where potential mistakes in the experiment can be identified.  

```{r set working directory or read in .CEL files}
# setting working directory and loading
setwd("~//Documents//GitHub//2022-topic-04-team-03//Data//rawData")

data.human=ReadAffy()
data.human@cdfName <- "HGU133Plus2_Hs_ENST"

```

### 3.2.1 Single chip control
The first step of QC is the single-chip analysis where major quality problems like fingerprints, local irregularities, imprints of pipette tips and extreme light intensities can be detected by a visual inspection of the microarray images. After looking at our chips we presupposed that the GSM456645 chip could be of bad quality due to a smudge on the bottom edge.

```{r chips pictures, eval =FALSE}
for (i in 1:18){
   image(data.human[,i], col = rainbow (100, start = 0, end = 0.75)[100:1],main=substr(colnames(data.human)[i],1,9)) }
```

```{r only chip GSM456645, echo=TRUE, include=TRUE, fig.cap="Figure 1: Single chip image of chip GSM456645"}
image(data.human[,3], col = rainbow (100, start = 0, end = 0.75)[100:1],main=substr(colnames(data.human)[3],1,9))
```
 
### Normalization

```{r normalization, echo=FALSE, message=FALSE, comment=FALSE, results="hide", highlight=TRUE}
#Normalization

human.vsnrma <- vsnrma(data.human)
setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")
colnames(human.vsnrma)= substr(colnames(human.vsnrma), 1,9)
expression.data <- exprs(human.vsnrma)
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
```


```{r meanSdPlot, echo=FALSE, message=FALSE, comment=FALSE, results="hide", eval=FALSE}
meanSdPlot(human.vsnrma)

setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//MeanSdPlot")
dev.copy2eps(file="meanSdPlot_human_vsnrma_normalized.eps")
##The line is not approximately horizontal, but like in project proposal said, so there shouldn't be a variance-mean dependence 
```

### Boxplots

```{r boxplots, eval =FALSE}
names = c("1 cell stage, rep 1", "1 cell stage, rep 2", "1 cell stage, rep 3",
          "2 cell stage, rep 1", "2 cell stage, rep 2", "2 cell stage, rep 3",
          "4 cell stage, rep 1", "4 cell stage, rep 2", "4 cell stage, rep 3",
          "8 cell stage, rep 1", "8 cell stage, rep 2", "8 cell stage, rep 3",
          "morula , rep 1", "morula , rep 2", "morula , rep 3",
          "blastocyst , rep 1", "blastocyst , rep 2", "blastocyst , rep 3")


# Before normalization:
par(las=2)
mmi=c(1,0.7,1.0477939,0.5366749)
par(mai=mmi)
boxplot(data.human, col= rainbow(35), cex.axis=0.5, main="Gene expression in human embroyogenesis data before normalization", names =names,las=2)
title(ylab="relative expression", line=2)
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//Boxplot")
dev.copy2eps(file="boxplot_rawdata_18290.eps")

# After Normalization:
boxplot(exprs(human.vsnrma), col= rainbow(35), cex.axis=0.5, main="Gene expression in human embroyogenesis data after normalization",names =names,las=2)
title(ylab="relative expression", line=2)
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//Boxplot")
dev.copy2eps(file="boxplot_normalizeddata_18290.eps")
```

### Density Plots

```{r density plots, eval =FALSE}
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//Density")
# Before Normalization:
hist(data.human, col=rainbow(35), main="Density function of log Intensity of human embryogenesis raw data")


dev.copy2pdf(file="Histogram_rawdata_39897.pdf", width = 10, height = 7)

# After Normalization:
eset=exprs(human.vsnrma)

plot(density(eset[,1]), type="n", xlab="log Intensity", ylim=c(0,1), main="Density function of log Intensity of human embryogenesis normalized")
for (i in 1:ncol(eset)) {
  lines(density(eset[,i]), col=rainbow(35)[i])
}

dev.copy2pdf(file="Histogram_NormalizedData_39897.pdf", width = 10, height = 7)
```

### RNA Degradation Plot

```{r RNA Degradation Plot, eval =FALSE}
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//RNAdegrad")
rnadeg.raw = AffyRNAdeg(data.human)

plotAffyRNAdeg(rnadeg.raw, col=rainbow(35))
title(sub="human embroyogenesis rawdata")
dev.copy2pdf(file="RNAdegrad_plot.pdf", width = 12.5, height = 20)
```


```{r RNA Degradation Plot shifted, eval =FALSE}
setwd("~//Documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//RNAdegrad")
plotAffyRNAdeg(rnadeg.raw, col=rainbow(35), transform= "shift.only")
title(sub="human embryogenesis rawdata - shifted")

dev.copy2pdf(file="RNAdegrad_plot_shifted.pdf", width = 12.5, height = 20)
```

### Scatter Plot

```{r scatter plot, eval =FALSE}
expression.data <- exprs(human.vsnrma)

for(i in 1:17){
  plot(expression.data[,c(i,i+1)], pch=".", cex=2)
  abline(0, 1, col="red")               # 45 degree dividing line
  
  title(main = paste("Scatterplot of probe", 
                     substr(colnames(human.vsnrma)[i], 1, nchar(colnames(human.vsnrma)[i])), "and", 
                     substr(colnames(human.vsnrma)[i+1], 1, nchar(colnames(human.vsnrma)[i+1])), 
                     sep=" ", collapse = NULL))
  
  file.name <- paste("~//documents//GitHub//2022-topic-04-team-03//Quality-Control//Plots//Scatterplot", 
                     as.character(substr(colnames(human.vsnrma)[i], 1, nchar(colnames(human.vsnrma)[i]))), "_",
                     as.character(substr(colnames(human.vsnrma)[i+1], 1, nchar(colnames(human.vsnrma)[i+1]))),
                     ".pdf", sep="")

  dev.copy2pdf(file = file.name)
}
```


### Data Clean-up

```{r data clean-up}
# Create a data frame out of the expression data from the normalized data
human.vsnrma.df = data.frame(exprs(human.vsnrma))

# Are there any NAs in our dataset?
sum(apply(human.vsnrma.df,1,function(x){sum(is.na(x))}))
#> No

```

## 3.2 Annotation

We want to annotate all the TRA genes in our chips with TRA and ensembl gene information

```{r Ensembl data}
# read in ensembl table with following attributes: "Gene stable ID", 
#"Gene stable ID version", "Transcript stable ID", "Transscript stable ID version", 
#"Gene.name", "Transcript name", "Chromosome.scaffold.name", "Gene.description", "HGNC.symbol"

setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")

ensembl.data <- read.csv("ensembl.human.txt")
dim(ensembl.data)
##268341      9

# create variables that contain gene ID, transcript ID, 
#the chromosome name and the gene symbol
ensembl.genes = ensembl.data[,1]
ensembl.transcripts = ensembl.data[,3]
ensembl.chromosome = ensembl.data[,7]
ensembl.symbols = ensembl.data[,9]
```



```{r clean human.vsnrma.df}
#rename the colnames to the stages
names = c("1 cell stage, rep 1", "1 cell stage, rep 2", "1 cell stage, rep 3",
          "2 cell stage, rep 1", "2 cell stage, rep 2", "2 cell stage, rep 3",
          "4 cell stage, rep 1", "4 cell stage, rep 2", "4 cell stage, rep 3",
          "8 cell stage, rep 1", "8 cell stage, rep 2", "8 cell stage, rep 3",
          "morula stage, rep 1", "morula stage, rep 2", "morula stage, rep 3",
          "blastocyst stage, rep 1", "blastocyst stage, rep 2", "blastocyst stage, rep 3")
colnames(human.vsnrma.df) = names

# Check dimensions
dim(human.vsnrma.df)
#95,721    18

# exclude Affymetrix control genes which begin with "AFFX"
human.vsnrma.df2 = human.vsnrma.df[63:95721,]
dim(human.vsnrma.df2)
# 95,659    18

# remove ".xx_at" suffix
rownames(human.vsnrma.df2) = substr(rownames(human.vsnrma.df2), 1,15)
```


```{r TRA data}
# read in TRA data for human
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")

tra.data <- read.table("tra.2017.human.gtex.5x.table.tsv",header=TRUE)
dim(tra.data)
#60131    10

# find the index of rownames of our dataset that contain transcript Ids from the tra dataset
# We extracted the TRA genes out of the three tables(human.vsnrma, tra, ensembl)
#TRAs in our dataset
j = which(rownames(human.vsnrma.df2) %in% tra.data$ensembl.transcript) 
tra.extracted = rownames(human.vsnrma.df2)[j] #24,783
human.vsnrma.only.tra = human.vsnrma.df2[j,]
dim(human.vsnrma.only.tra)
#24783 18 

#exclude TRAs which are not expressed in our dataset
k = which(tra.data$ensembl.transcript %in% tra.extracted)
tra.chips = tra.data[k,] #24,783 
dim(tra.chips)
#24783 10
```



```{r Extract TRAs in Ensembl }
#we want to now extract the genes from Ensembl table that are present in the TRAs expressed in our chips
c = which(ensembl.transcripts %in% tra.extracted)
ensembl.only.tra = ensembl.data[c,]
dim(ensembl.only.tra)
#24623 9
```

There are 160 genes which are in the TRA table but not in Ensembl table. We downloaded and checked the Ensembl table again, which did not help.


```{r TRA Annotation}
#reorder the rows of ensembl.only.tra and tra.chips
tra.chips = arrange(tra.chips,ensembl.transcript)
ensembl.only.tra = arrange(ensembl.only.tra,Transcript.stable.ID)

#fusion of two tables: tra.chips and human.vsnrma.only.tra
fusion.tra.chips = cbind(human.vsnrma.only.tra,tra.chips$ensembl.gene,tra.chips$tiss.number,tra.chips$tissues,tra.chips$max.tissue)
colnames(fusion.tra.chips)[19:22]<-c("ensembl.gene","tissue.number","tissues","max.tissue")

# fusion of fusion.tra.chips and ensembl.only.tra (aware that ensembl.only.tra table contains less rows than the other two tables due to the 160 missing genes)
p = which(rownames(fusion.tra.chips) %in% ensembl.transcripts)
tra.ensembl = fusion.tra.chips[p,]
annotated.tra = cbind(tra.ensembl,ensembl.only.tra$Gene.name,ensembl.only.tra$Gene.description,ensembl.only.tra$Chromosome.scaffold.name)

#rename colnames
colnames(annotated.tra)[23:25]<-c("Gene.name","Gene.description","Chromosome.scaffold.name")


setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
write.csv(annotated.tra, file="annotated.tra.csv")

```

### 3.3 Exploratory data analysis

We want to visualize the distribution of the TRAs in our chips according to their max tissue, using a boxplot.

```{r Tissue distribution plot}
# Generate a list containing all tissue names with number of tissue TRAs, which is detected in our chips
tissue=data.frame(table(tra.data$max.tissue))$Var1
length(tissue) #53
tissue.number = c()

for (i in 1:53) {
  tissue.number[[i]] = nrow(annotated.tra %>% filter(grepl(tissue[i],annotated.tra[,21])))
}

tissue.distribution = cbind(as.vector(tissue),tissue.number)
colnames(tissue.distribution)=c("max.tissue","tissue.number")
tissue.distribution = as.data.frame(tissue.distribution)

# Arranging the list according to the descending TRA number
tissue.distribution.arranged = arrange(tissue.distribution, desc(as.numeric(tissue.number)))

# Create a barplot to present the TRA distribution of our chips

par(mar=c(5,9,4,1)+.1)
barplot(height = as.numeric(tissue.distribution.arranged$tissue.number[1:10]),names.arg = tissue.distribution.arranged$max.tissue[1:10],cex.names=0.5,col=rainbow(10), main="Frequency of TRAs in our data", xlab="Frequency",las=2, horiz= TRUE, cex.lab=0.5, cex.axis= 0.5)


setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")

dev.copy2pdf(file="tissue.distribution.arranged.barplot.pdf" )
```

shitty heatamp, do we want to keep it?

```{r heatmap, eval =FALSE}
# Create a heatmap of the TRA genes in our dataset
## What does unlist mean? -> Transforms list into matrix
human.vsnrma.only.tra.matrix = matrix(unlist(human.vsnrma.only.tra), ncol = 18, nrow = 24783)

colnames(human.vsnrma.only.tra.matrix) = names
heatmap = pheatmap(human.vsnrma.only.tra.matrix, 
                   cluster_cols=FALSE, 
                   show_rownames = TRUE, 
                   legend=TRUE, 
                   fontsize_row=0.5,
                   main = "Expression of 24,783 TRA genes in 6 different stages of embryogenesis")

setwd("~//documents//GitHub//2022-topic-04-team-03//Plots")
dev.copy2pdf(file="heatmap.pdf")

```

## 3.4 Data analysis
### 3.4.1 Principle Component Analysis

Using PCA to analyse our 18 chips, to see whether there are any outlier-chip and whether the replicates from different stages cluster with each other.

```{r PCA}
#conducting PCA
topVar = apply(human.vsnrma.df2, 1, var)
q75 = quantile(topVar, probs = 0.75)
i.topvar = which(topVar >= q75)
human.vsnrma.df2.topVar = human.vsnrma.df2[i.topvar,]
dim(human.vsnrma.df2.topVar)

pca = prcomp(t(human.vsnrma.df2.topVar), center = T, scale. = T)

```

```{r show how much variance each PC explains}
#show how much variance each PC explains

variance = (pca$sdev)^2
prop.variance = variance/sum(variance)

# Plot cumulative proportion of variance explained
plot(cumsum(prop.variance), xlab = "Principal component",
     ylab = "Cumulative proportion of variance explained",cex.lab=0.7,
     ylim = c(0, 1), type = "b")

abline(h=0.9)
```

4 PCs are enough to show more than 90%

```{r Plot PC1 and PC2, echo=TRUE}
color1 = c(rep("red",3),rep("orange",3),rep("yellow",3),rep("green",3),rep("blue",3),rep("purple",3))

par(mar=c(5, 4, 4, 10), xpd=TRUE)

plot(pca$x[,1], pca$x[,2],col=color1, pch=19,xlab='PC1',ylab='PC2',main="Principal component analysis")

color2= c("red","orange","yellow","green","blue","purple")

legend(x = 190,y=130,legend = c("1-cell stage", "2-cell stage","4-cell stage", "8-cell stage", "Morula stage","Blastocyst stage"), fill= color2,col = color2)


```

We cannot observe any outlier replicate and we can see four cluster (Cluster 1:1-cell,2-cell and 4-cell stage, Cluster 2: 8-cell stage, Cluster 3:Morula, Cluster 4:Blastocyst)

```{r plots PC2/3 and PC3/4, eval=FALSE}
#Plot PC2 and PC3

par(mar=c(5, 4, 4, 10), xpd=TRUE)

plot(pca$x[,2], pca$x[,3],col=color1, pch=19,xlab='PC2',ylab='PC3',main="Principal component analysis")

legend(x = 140,y=92,legend = c("1-cell stage", "2-cell stage","4-cell stage", "8-cell stage", "Morula stage","Blastocyst stage"), fill= color2,col = color2)

#Plot PC3 and PC4
par(mar=c(5, 4, 4, 10), xpd=TRUE)

plot(pca$x[,3], pca$x[,4],col=color1, pch=19,xlab='PC3',ylab='PC4',main="Principal component analysis")

color2= c("red","orange","yellow","green","blue","purple")

legend(x = 100,y=48,legend = c("1-cell stage", "2-cell stage","4-cell stage", "8-cell stage", "Morula stage","Blastocyst stage"), fill= color2,col = color2)


```

Now we want to use Kmeans on the first 4 PCs (which explain 90% of the Variance) to find potential clusters and compare it with the cluster that we have observed in PCA just now

```{r run models with varying value of k}
#Use map_dbl to run many models with varying value of k (centers)
tot_withinss <- map_dbl(1:10,  function(k){
  model <- kmeans(x = pca$x[,1:4], centers = k,nstart=30)
  model$tot.withinss
})
```

```{r data frame with k & tot_withinns}
# Generate a data frame containing both k and tot_withinss
elbow_df <- data.frame(
  k = 1:10 ,
  tot_withinss = tot_withinss
)
```

```{r plot with elbow plot, echo=TRUE}
ggplot(elbow_df, aes(x = k, y =tot_withinss)) +
  geom_line() +
  scale_x_continuous(breaks = 1:10) + labs(y= "Total within sum of squares")+ ggtitle("Elbow plot")
```

seems like k=2 is the best

```{r show cluster, echo=TRUE}
model <- kmeans(x = pca$x[,1:4], centers = 2,nstart=30)
b<- as.data.frame(model$cluster)
colnames(b) <- "Cluster"
b
```

Elbow plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)


Plot the Siloutte plot

```{r Use map_dbl to run many models with varying value of k}
sil_width <- map_dbl(2:10,  function(k){
  model <- pam(pca$x[,1:4], k = k)
  model$silinfo$avg.width
})
```

```{r Silhouette plot, echo =TRUE}
# Generate a data frame containing both k and sil_width
sil_df <- data.frame(
  k = 2:10,
  sil_width = sil_width
)

# Plot the relationship between k and sil_width
ggplot(sil_df, aes(x = k, y = sil_width)) +
  geom_line() +
  scale_x_continuous(breaks = 2:10) + labs(y= "Silhouette Width")+ ggtitle("Silhouette plot")
```

it seems like that k=4 is a good choice


```{r Cluster silhouette plot, echo=TRUE }
# Generate a k-means model using the pam() function with a k = 4
pam_k4 <- pam(pca$x[,1:4],k=4)

# Show the clustering
a <-as.data.frame(pam_k4$clustering)
colnames(a)= "Cluster"
a
```

Silouette Plot suggests 4 Clusters: Cluster 1:1-cell,2-cell and 4-cell stage, Cluster 2: 8-cell stage, Cluster 3:Morula, Cluster 4:Blastocyst)


Using Kmeans based on the original expression data (human.vsnrma.df2) to cluster and compare the results with the Kmeans results based on 4PCs

```{r kmeans on whole data}
# Use map_dbl to run many models with varying value of k (centers)
tot_withinss <- map_dbl(1:10,  function(k){
    model <- kmeans(x = t(human.vsnrma.df2), centers = k,nstart=30)
    model$tot.withinss
})
```

```{r Elbow plot}
# Generate a data frame containing both k and tot_withinss
elbow_df <- data.frame(
    k = 1:10 ,
    tot_withinss = tot_withinss
)

# Plot the elbow plot
ggplot(elbow_df, aes(x = k, y =tot_withinss)) +
    geom_line() +
    scale_x_continuous(breaks = 1:10)+ labs(y= "Total within sum of squares")+ ggtitle("Elbow plot")
```

seems like k=2 is the best

```{r}
#show cluster
model <- kmeans(x = t(human.vsnrma.df2), centers = 2,nstart=30)
c<-as.data.frame(model$cluster)
colnames(c)="Cluster"
c
```

Elbow plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)


Plot the Silhoutte plot

```{r Silhoutte plot whole data}
# Use map_dbl to run many models with varying value of k
sil_width <- map_dbl(2:10,  function(k){
    model <- pam(t(human.vsnrma.df2), k = k)
    model$silinfo$avg.width
})

# Generate a data frame containing both k and sil_width
sil_df <- data.frame(
    k = 2:10,
    sil_width = sil_width
)

# Plot the relationship between k and sil_width
ggplot(sil_df, aes(x = k, y = sil_width)) +
    geom_line() +
    scale_x_continuous(breaks = 2:10)+ labs(y= "Silhouette Width")+ ggtitle("Silhouette plot")
```

it seems like that k=2 is a good choice

```{r}
# Generate a k-means model using the pam() function with a k = 4
pam_k2 <- pam(t(human.vsnrma.df2),k=2)
  
# Show the clustering
d<- as.data.frame(pam_k2$clustering)
colnames(d) <- "Cluster"
d
```

Silouette plot suggests 2 Clusters (Cluster 1: 1-cell stage, 2-cell stage and 4-cell stage, Cluster 2: 8-cell stage, morula and blastocyst)

Finding genes which explain for the variance by looking at genes with the highst loading in PCs

Use PCA to plot 3 replicates of 1-cell stage and 8-cell stage

```{r}
a=c(1,2,3,10,11,12)
pca.1.8=prcomp(t(human.vsnrma.df2[,a]))
```

```{r}
# Plot cumulative proportion of variance explained
variance = (pca.1.8$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
```

two PCs are enough to show 90% variance

Finding variance genes which explain the variance of PC1 between 1-cell stage and 8-cell stage

```{r}
pc1=pca.1.8$rotation[,1]
top_genes_pca1 = names(sort(abs(pc1), decreasing = T)[1:10])

#Variance genes 1-cell stage and 8-cell stage
j = which(ensembl.data$Transcript.stable.ID%in% top_genes_pca1)

#Variance TRA genes 1-cell stage and 8-cell stage
k = which(rownames(annotated.tra)%in% top_genes_pca1)
annotated.tra[k,]
```

```{r}
#Use PCA to plot 3 replicates of 8-cell stage and morula stage
pca.8.m=prcomp(t(human.vsnrma.df2[,10:15]))
```

```{r}
# Plot cumulative proportion of variance explained
plot(pca.8.m$sdev)
variance = (pca.8.m$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
```

three PCs are enough to show 90% variance

Finding variance genes which explain the variance of PC1 between 8-cell stage and morula stage

```{r}
pc1=pca.8.m$rotation[,1]
top_genes_pca1 = names(sort(abs(pc1), decreasing = T)[1:10])

#Variance genes 8-cell stage and morula stage
j = which(ensembl.data$Transcript.stable.ID%in% top_genes_pca1)


#Variance TRA genes 8-cell stage and morula stage
k = which(rownames(annotated.tra)%in% top_genes_pca1)
annotated.tra[k,]
```

*Use PCA to plot 3 replicates of 8-cell stage and blastocyst stage*

Use PCA to plot 3 replicates of morula stage and blastocyst stage

```{r}
pca.m.b=prcomp(t(human.vsnrma.df2[,13:18]))

# Plot cumulative proportion of variance explained
plot(pca.m.b$sdev)
variance = (pca.m.b$sdev)^2
prop.variance = variance/sum(variance)
names(prop.variance) = 1:length(prop.variance)
barplot(prop.variance[1:20],ylab='Proportion of variance') # we only plot the first 20 PCs
plot(cumsum(prop.variance), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     ylim = c(0, 1), type = "b")

abline(h=0.9)
```

four PCs are enough to show 90% variance

Finding variance genes which explain the variance of PC1 between morula stage and blastocyst stage

```{r}
pc1=pca.m.b$rotation[,1]
top_genes_pca1 = names(sort(abs(pc1), decreasing = T)[1:10])

#Variance genes morula stage and blastocyst stage
j = which(ensembl.data$Transcript.stable.ID%in% top_genes_pca1)
ensembl.data[j,]

#Variance tra genes morula stage and blastocyst stage
k = which(rownames(annotated.tra)%in% top_genes_pca1)
annotated.tra[k,]
```

### 3.4.2 Limma analysis

Using Limma analysis, we want to find out differential expressed genes (DEGs) between different stages. We will conduct the test among every two stages but only look at the results of the stages between different clusters (showed in PCA): stage 1 cell-8 cell, stage 8 cell - morula, stage morula - blastocyst

```{r limma Analysis}

# Define the different stages of the chips in a vector
stage = factor(x= c(1,1,1,2,2,2,4,4,4,8,8,8,"morula","morula","morula", "blastocyst", "blastocyst", "blastocyst"),levels= c(1,2,4,8,"morula","blastocyst"))

# Create a design matrix with the stages
design = model.matrix(~0 + stage)
            
# Create a contrast matrix which compares all the stages with one another 
cm = makeContrasts(stage1.2 = stage2-stage1, stage1.4 = stage4-stage1, stage1.8 = stage8-stage1, stage1.morula = stagemorula-stage1, stage1.blastocyst =stageblastocyst-stage1,
                   stage2.4 = stage4-stage2, stage2.8 = stage8-stage2, stage2.morula = stagemorula-stage2,  stage2.blastocyst = stageblastocyst-stage2,
                   stage4.8 = stage8-stage4, stage4.morula = stagemorula-stage4,  stage4.blastocyst = stageblastocyst-stage4,
                   stage8.morula = stagemorula-stage8, stage8.blastocyst = stageblastocyst-stage8,
                   stagemorula.blastocyst =stageblastocyst -stagemorula, 
                   levels = design)

# Fit the coefficients of the model
fit = lmFit(human.vsnrma.df2, design)
fit2 = contrasts.fit(fit, contrasts = cm)

# Calculate the t-statistics
fit2 = eBayes(fit2)

# Number of differentially expressed genes (p-value = 0.05)
results = decideTests(fit2, p.value = 0.05)
summary(results)

# Number of differentially expressed genes (p-value = 0.01)
results2 = decideTests(fit2, p.value = 0.01)
summary(results2)
```
*creating fit tables between every two stages*

```{r fitFunction}
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")

fit.function = function(i){
  fit.table = contrasts.fit(fit, contrasts = cm[,i])
  fit.table = eBayes(fit.table)
}
```

```{r limma fit tables}
fit.1.2 = fit.function(1)
fit.1.4 = fit.function(2)
fit.1.8 = fit.function(3)
fit.1.m = fit.function(4)
fit.1.b = fit.function(5)
fit.2.4 = fit.function(6)
fit.2.8 = fit.function(7)
fit.2.m = fit.function(8)
fit.2.b = fit.function(9)
fit.4.8 = fit.function(10)
fit.4.m = fit.function(11)
fit.4.b = fit.function(12)
fit.8.m = fit.function(13)
fit.8.b = fit.function(14)
fit.m.b = fit.function(15)
```

*using toptable function to create limma table (only significant DEGs!) for every two stages*

```{r topTable function}
top.table = function(i){
  fit.table = contrasts.fit(fit, contrasts = cm[,i])
  fit.table = eBayes(fit.table)
  
  pvalue01 = sum(p.adjust(fit.table$p.value, "BH") < 0.01)
  topTable(fit.table, number = pvalue01)
}
```

```{r limma topTables}
limma.table.1.2 = top.table(1)
limma.table.1.4 = top.table(2)
limma.table.1.8 = top.table(3)
limma.table.1.m = top.table(4)
limma.table.1.b = top.table(5)
limma.table.2.4 = top.table(6)
limma.table.2.8 = top.table(7)
limma.table.2.m = top.table(8)
limma.table.2.b = top.table(9)
limma.table.4.8 = top.table(10)
limma.table.4.m = top.table(11)
limma.table.4.b = top.table(12)
limma.table.8.m = top.table(13)
limma.table.8.b = top.table(14)
limma.table.m.b = top.table(15)
```

*we annotate the limma table (only significant DEGs) with ensembl*

```{r annotation function}
limma.annotation = function(x){
a = which(ensembl.transcripts %in% rownames(x))
ensembl.new = ensembl.data[a,]
  
b = which(rownames(x) %in% ensembl.new$Transcript.stable.ID)
x.new = x[b,]
  
ensembl.new = arrange(ensembl.new,Transcript.stable.ID)
x.new = arrange(x.new, rownames(x.new))
  
annotated.x = cbind(x.new, ensembl.new$Gene.name, ensembl.new$Gene.description,ensembl.new$Chromosome.scaffold.name, ensembl.new$Gene.stable.ID)
colnames(annotated.x)[7:10]<-c("gene.name","gene.description","Chromosome.scaffold.name", "Gene.stable.ID")
annotated.x = arrange(annotated.x, adj.P.Val)
  return(annotated.x)
}
```

```{r limma annotation}
annotated.limma.1.8 = limma.annotation(limma.table.1.8)
annotated.limma.1.m = limma.annotation(limma.table.1.m)
annotated.limma.1.b = limma.annotation(limma.table.1.b)
#annotated.limma.2.4 = limma.annotation(limma.table.2.4)
annotated.limma.2.8 = limma.annotation(limma.table.2.8)
annotated.limma.2.m = limma.annotation(limma.table.2.m)
annotated.limma.2.b = limma.annotation(limma.table.2.b)
annotated.limma.4.8 = limma.annotation(limma.table.4.8)
annotated.limma.4.m = limma.annotation(limma.table.4.m)
annotated.limma.4.b = limma.annotation(limma.table.4.b)
annotated.limma.8.m = limma.annotation(limma.table.8.m)
annotated.limma.8.b = limma.annotation(limma.table.8.b)
annotated.limma.m.b = limma.annotation(limma.table.m.b)
```

*we extract the TRAs from ensembl annotated limma table (only significant DEGs) and annotate these significant differentially expressed TRAs with TRA information*

```{r limma TRAs}
#annotated limma tables mit TRAs annotaten
#1.4 und 1.2 sind 00
nr<-c(1,9,10)

annotated.limma.1.8$ensembl.transcript = rownames(annotated.limma.1.8)
annotated.limma.1.8.tra <- merge(annotated.limma.1.8, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.1.m$ensembl.transcript = rownames(annotated.limma.1.m)
annotated.limma.1.m.tra <- merge(annotated.limma.1.m, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.1.b$ensembl.transcript = rownames(annotated.limma.1.b)
annotated.limma.1.b.tra <- merge(annotated.limma.1.b, tra.data[,nr], by = 'ensembl.transcript')
#2.4. ist 00
annotated.limma.2.8$ensembl.transcript = rownames(annotated.limma.2.8)
annotated.limma.2.8.tra <- merge(annotated.limma.2.8, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.2.m$ensembl.transcript = rownames(annotated.limma.2.m)
annotated.limma.2.m.tra <- merge(annotated.limma.2.m, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.2.b$ensembl.transcript = rownames(annotated.limma.2.b)
annotated.limma.2.b.tra <- merge(annotated.limma.2.b, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.4.8$ensembl.transcript = rownames(annotated.limma.4.8)
annotated.limma.4.8.tra <- merge(annotated.limma.4.8, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.4.m$ensembl.transcript = rownames(annotated.limma.4.m)
annotated.limma.4.m.tra <- merge(annotated.limma.4.m, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.4.b$ensembl.transcript = rownames(annotated.limma.4.b)
annotated.limma.4.b.tra <- merge(annotated.limma.4.b, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.8.m$ensembl.transcript = rownames(annotated.limma.8.m)
annotated.limma.8.m.tra <- merge(annotated.limma.8.m, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.8.b$ensembl.transcript = rownames(annotated.limma.8.b)
annotated.limma.8.b.tra <- merge(annotated.limma.8.b, tra.data[,nr], by = 'ensembl.transcript')

annotated.limma.m.b$ensembl.transcript = rownames(annotated.limma.m.b)
annotated.limma.m.b.tra <- merge(annotated.limma.m.b, tra.data[,nr], by = 'ensembl.transcript')

```

#### Tissue plot

we want to visualize the expression of TRAs among the different stages

```{r tissue plot}
#Erstellen Means pro stage pro tissue
n_occur1.8=data.frame(table(annotated.limma.1.8.tra$max.tissue))
Δ1.8=c()
for (i in 1:53) {

  tissue.genes=which(annotated.limma.1.8.tra$max.tissue %in% n_occur1.8$Var1[i])
   Δ1.8[[i]]= mean(c(annotated.limma.1.8.tra[tissue.genes,2]))
}


n_occur1.m=data.frame(table(annotated.limma.1.m.tra$max.tissue))
Δ1.m=c()
for (i in 1:53) {

  tissue.genes=which(annotated.limma.1.m.tra$max.tissue %in% n_occur1.m$Var1[i])
  Δ1.m[[i]]= mean(c(annotated.limma.1.m.tra[tissue.genes,2]))
}

n_occur1.b=data.frame(table(annotated.limma.1.b.tra$max.tissue))
Δ1.b=c()
for (i in 1:53) {

  tissue.genes=which(annotated.limma.1.b.tra$max.tissue %in% n_occur1.b$Var1[i])
   Δ1.b[[i]]= mean(c(annotated.limma.1.b.tra[tissue.genes,2]))
}
Δ1.2=c()
Δ1.2[1:53] =0
Δ1.4=c()
Δ1.4[1:53] =0
#Erstellen Plot Data
tissue.matrix = cbind(Δ1.2,Δ1.4,Δ1.8,Δ1.m,Δ1.b)
rownames(tissue.matrix) = n_occur1.8$Var1[1:53] 
tissue.matrix= cbind(rownames(tissue.matrix),tissue.matrix)
tissue.matrix= as_tibble(tissue.matrix)


tissue.matrix=pivot_longer(tissue.matrix, cols=Δ1.2:Δ1.b, names_to = "stages" ,values_to="logFc")

#plot
ggplot(tissue.matrix, aes(x = stages, y= as.numeric(logFc),group=as.character(V1)))  + geom_line(aes(color=as.character(V1))) +
           geom_point(aes(color=as.character(V1))) + labs(y=("logFC")) + labs(colour = "tissues")

```

###tissue plot facetting

```{r facetting}
tissue.max.vector = c(rep("Adipose",10),rep("Adrenal Gland",5),rep("Artery",15),rep("Bladder",5),rep("Brain",65),rep("Breast",5),rep("Cells",10),rep("Cervix",10),rep("Colon",10),rep("Esophagus", 15),rep("Follopian Tube",5),rep("Heart",10),rep("Kidney",5),rep("Liver",5),rep("Lung",5),rep("Minor Salivary Gland",5),rep("Muscle",5),rep("Nerve",5),rep("Ovary",5),rep("Pancreas",5),rep("Pituitary",5),rep("Prostate",5),rep("Skin",10),rep("Small Intestine",5), rep("Spleen",5), rep("Stomach",5),rep("Testis",5),rep("Thyroid",5),rep("Uterus",5),rep("Vagina",5),rep("Whole Blood",5))

tissue.fac= cbind(tissue.matrix,tissue.max.vector)
plot.tissue.fac <- ggplot(tissue.fac, aes(x = stages, y= as.numeric(logFc),group=as.character(V1)))  + geom_line(aes(color=as.character(V1))) +
           geom_point(aes(color=as.character(V1))) + labs(y=("logFC")) + labs(colour = "tissues") +facet_wrap(~tissue.max.vector)#+ theme(legend.position = "none")
plot.tissue.fac
#uncomment last component to remove legend 
```

#### filter DE genes & volcanoplots

we create volcanoplots to visualize the DEGs

so far, our limma table only contains significant DEGs. To demonstrate other genes that do not show differential expression we need to create a limma table with all genes. By doing that we found that the ensembl table has some transcripts that repeats itself, so we only select the unique transcripts from the ensembl data

```{r}
#find out repeating transcripts in ensembl.data
n_occur=data.frame(table(ensembl.data$Transcript.stable.ID))
n_occur=arrange(n_occur,desc(n_occur$Freq))
a=sum(n_occur$Freq>1)
duplicate.ensembl=n_occur$Var1[1:40]
j=which(ensembl.data$Transcript.stable.ID%in%duplicate.ensembl)  
ensembl.duplicate=ensembl.data[j,]
ensembl.duplicate=arrange(ensembl.duplicate,ensembl.duplicate$Transcript.stable.ID)

#delete the repeating transcripts
Nth.delete<-function(dataframe, n)dataframe[-(seq(n,to=nrow(dataframe),by=n)),]
ensembl.duplicate.genes=Nth.delete(ensembl.duplicate, 2)
ensembl.unique=rbind(ensembl.data[-j,],ensembl.duplicate.genes)
```


generate a function with the output of limma tables with sig. genes and non sig. genes

```{r limma tables with sig & non sig. genes}
generate.limma.table.vollst = function(fit,n){
  fit.1 = contrasts.fit(fit, contrasts = cm[,n])
  fit.1 = eBayes(fit.1)
  limma.table.vollst=topTable(fit.1,number = dim(fit.1)[1])
  return(limma.table.vollst)
}
limma.table.vollst.1.2=generate.limma.table.vollst(fit,1)
limma.table.vollst.1.4=generate.limma.table.vollst(fit,2)
limma.table.vollst.1.8=generate.limma.table.vollst(fit,3)
limma.table.vollst.1.m=generate.limma.table.vollst(fit,4)
limma.table.vollst.1.b=generate.limma.table.vollst(fit,5)
limma.table.vollst.2.4=generate.limma.table.vollst(fit,6)
limma.table.vollst.2.8=generate.limma.table.vollst(fit,7)
limma.table.vollst.2.m=generate.limma.table.vollst(fit,8)
limma.table.vollst.2.b=generate.limma.table.vollst(fit,9)
limma.table.vollst.4.8=generate.limma.table.vollst(fit,10)
limma.table.vollst.4.m=generate.limma.table.vollst(fit,11)
limma.table.vollst.4.b=generate.limma.table.vollst(fit,12)
limma.table.vollst.8.m=generate.limma.table.vollst(fit,13)
limma.table.vollst.8.b=generate.limma.table.vollst(fit,14)
limma.table.vollst.m.b=generate.limma.table.vollst(fit,15)
```

anotation of the limma table (complete) with emsembl information
```{r Annotation of limma including all genes}

ensembl.limma.annotation <- function(x){
  a = which(ensembl.unique$Transcript.stable.ID %in% rownames(x))
  ensembl.new = ensembl.unique[a,]
  
  b = which(rownames(x)%in% ensembl.new$Transcript.stable.ID)
  x.new = x[b,]
  
  ensembl.new = arrange(ensembl.new,ensembl.new$Transcript.stable.ID)
  x.new = arrange(x.new, rownames(x.new))
  
  annotated.x = cbind(x.new, ensembl.new$Gene.name, ensembl.new$Gene.description,ensembl.new$Chromosome.scaffold.name)
  colnames(annotated.x)[7:9]<-c("gene.name","gene.description","Chromosome.scaffold.name")
  annotated.x = arrange(annotated.x, adj.P.Val)
  return(annotated.x)
}

annotated.limma.table.vollst.1.8 = ensembl.limma.annotation(limma.table.vollst.1.8)
annotated.limma.table.vollst.8.m = ensembl.limma.annotation(limma.table.vollst.8.m) 
annotated.limma.table.vollst.m.b = ensembl.limma.annotation(limma.table.vollst.m.b)
```

we want to plot three volcanoplots (between 1-cell and 8-cell stage, 8-cell stage and morula stage, morula stage and blastocyst stage). We decided to highlight 10 highst up-regulated TRAs and 10 highst down-regulated TRAs.

```{r volcano plots 1.8, echo=TRUE}
#group the genes into "Up-regulated","Down-regulated" and "Unchanged" for colouring volcano plot. We definded a gene as up/down-regulated wenn the LogFC is higher than 2 or smaller than -2 

#for the stage between 1-cell and 8-cell stage
annotated.limma.table.vollst.1.8 <- annotated.limma.table.vollst.1.8 %>% 
  mutate(
    Expression = case_when(logFC > 2 & adj.P.Val < 0.01 ~ "Up-regulated",
                           logFC < -2 & adj.P.Val < 0.01 ~ "Down-regulated",
                           TRUE ~ "not significantly changed"))


#extract 10 highst up-regulated TRAs and 10 highst down-regulated TRAs arranged by both logFC and adj.P.Value

top <- 20
top_genes_tra.1.8 <- bind_rows(
  annotated.limma.1.8.tra %>% 
    filter(logFC >2) %>% 
    arrange(adj.P.Val,desc(abs(logFC))) %>% 
    head(top),
  annotated.limma.1.8.tra %>% 
    filter(logFC< -2) %>% 
    arrange(adj.P.Val,desc(abs(logFC)))
  %>%head(top)
  )
top_genes_tra.1.8
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
write.csv(top_genes_tra.1.8, file="top_genes_tra.1.8.csv")

#create a volcano plot
p1 <- ggplot(annotated.limma.table.vollst.1.8, aes(logFC, -log(adj.P.Val,10)))+
  geom_point(aes(color = Expression),size = 2/5)+  
  ggtitle("Volcano plot between 1-cell stage and 8-cell stage")+
  theme(plot.title = element_text(hjust = 0.5,face = "bold",size=14))+
  xlab(expression("log"[10]*"FC")) + 
  ylab(expression("-log"[10]*"(adj.P.Val)"))+
  geom_vline(xintercept=c(-2, 2), col="black",linetype = "dashed") +
  geom_hline(yintercept=-log10(0.01), col="black",linetype = "dashed")

#label the 20 most interesting differencially expressed TRA genes
p1 <-  p1 +
  geom_label_repel(data = top_genes_tra.1.8,
                   mapping = aes(logFC, -log(adj.P.Val,10), label = gene.name),
                   size = 1.5)
p1
```
Repeating the same procedure with the other three stages
```{r Volcano plot 8.m, echo=TRUE}
#group the genes into "Up-regulated","Down-regulated" and "Unchanged" for colouring volcano plot
annotated.limma.table.vollst.8.m <- annotated.limma.table.vollst.8.m %>% 
  mutate(
    Expression = case_when(logFC >= 2 & adj.P.Val < 0.01 ~ "Up-regulated",
                           logFC <= -2 & adj.P.Val < 0.01 ~ "Down-regulated",
                           TRUE ~ "not significantly changed"))


#extract 10 highst up-regulated TRAs and 10 highst down-regulated TRAs arranged by both logFC and adj.P.Value

top <- 10
top_genes_tra.8.m <- bind_rows(
  annotated.limma.8.m.tra %>% 
    filter(logFC >2) %>% 
    arrange(adj.P.Val,desc(abs(logFC))) %>% 
    head(top),
  annotated.limma.8.m.tra %>% 
    filter(logFC< -2) %>% 
    arrange(adj.P.Val,desc(abs(logFC)))
  %>%head(top)
)
top_genes_tra.8.m
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
write.csv(top_genes_tra.8.m, file="top_genes_tra.8.m.csv")


#create a volcanoplot
p2 <- ggplot(annotated.limma.table.vollst.8.m, aes(logFC, -log(adj.P.Val,10)))+
  geom_point(aes(color = Expression),size = 2/5)+  
  ggtitle("Volcano plot between 8-cell stage and morula stage")+
  theme(plot.title = element_text(hjust = 0.5,face = "bold",size=10))+
  xlab(expression("log"[10]*"FC")) + 
  ylab(expression("-log"[10]*"(adj.P.Val)"))+
  geom_vline(xintercept=c(-2, 2), col="black",linetype = "dashed") +
  geom_hline(yintercept=-log10(0.01), col="black",linetype = "dashed")

#label the 20 most interesting differentially expressed TRA genes
p2 <-  p2 +
  geom_label_repel(data = top_genes_tra.8.m,
                   mapping = aes(logFC, -log(adj.P.Val,10), label = gene.name),
                   size = 1.5)
p2
```

```{r Volcano plot m.b ,echo=TRUE}
#group the genes into "Up-regulated","Down-regulated" and "Unchanged" for colouring volcano plot
annotated.limma.table.vollst.m.b <- annotated.limma.table.vollst.m.b %>% 
  mutate(
    Expression = case_when(logFC >= 2 & adj.P.Val < 0.01 ~ "Up-regulated",
                           logFC <= -2 & adj.P.Val < 0.01 ~ "Down-regulated",
                           TRUE ~ "not significantly changed"))


#extract 10 highst up-regulated TRAs and 10 highst down-regulated TRAs arranged by both logFC and adj.P.Value

top <- 10
top_genes_tra.m.b <- bind_rows(
  annotated.limma.m.b.tra %>% 
    filter(logFC >2) %>% 
    arrange(adj.P.Val,desc(abs(logFC))) %>% 
    head(top),
  annotated.limma.m.b.tra %>% 
    filter(logFC < -2) %>% 
    arrange(adj.P.Val,desc(abs(logFC)))
  %>%head(top)
)
top_genes_tra.m.b
setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
write.csv(top_genes_tra.m.b, file="top_genes_tra.m.b.csv")


#create a volcanoplot
p3 <- ggplot(annotated.limma.table.vollst.m.b, aes(logFC, -log(adj.P.Val,10)))+
  geom_point(aes(color = Expression),size = 2/5)+  
  ggtitle("Volcanoplot between morula stage and blastocyst stage")+
  theme(plot.title = element_text(hjust = 0.5,face = "bold",size=10))+
  xlab(expression("log"[10]*"FC")) + 
  ylab(expression("-log"[10]*"(adj.P.Val)"))+
  geom_vline(xintercept=c(-2, 2), col="black",linetype = "dashed") +
  geom_hline(yintercept=-log10(0.01), col="black",linetype = "dashed")

#label the 20 most interesting differentially expressed TRA genes
p3 <-  p3 +
  geom_label_repel(data = top_genes_tra.m.b,
                   mapping = aes(logFC, -log(adj.P.Val,10), label = gene.name),
                   size = 1.5)
p3
```

#### Expression plots interesting genes

```{r Expression plot STMN1}
names1 = c("1 cell stage", 
          "2 cell stage",
          "4 cell stage",
          "8 cell stage",
          "morula stage",
          "blastocyst stage")


annotated.tra.mean= cbind(rownames(annotated.tra),rowMeans(annotated.tra[,1:3]),rowMeans(annotated.tra[,4:6]),rowMeans(annotated.tra[,7:9]),rowMeans(annotated.tra[,10:12]),rowMeans(annotated.tra[,13:15]),rowMeans(annotated.tra[,16:18]),annotated.tra[,19:25])

colnames(annotated.tra.mean)[2:7]=names1

o = c(which(annotated.tra.mean[,12] %in% "STMN1"))

annotated.tra.mean.1=pivot_longer(annotated.tra.mean[o,], cols=2:7, names_to = "Stages" ,values_to="expression")

colnames(annotated.tra.mean.1)[1]= "transcript"

ggplot(annotated.tra.mean.1, aes(x = Stages, y =expression,group=as.character(transcript))) +
  geom_line(aes(color=transcript)) + scale_x_discrete(limits=names1) +
  labs(y=("Expression")) +
    ggtitle("Expression STMN1")
```

```{r Expression plot SLC2A3}
names1 = c("1 cell stage", 
          "2 cell stage",
          "4 cell stage",
          "8 cell stage",
          "morula stage",
          "blastocyst stage")


annotated.tra.mean= cbind(rownames(annotated.tra),rowMeans(annotated.tra[,1:3]),rowMeans(annotated.tra[,4:6]),rowMeans(annotated.tra[,7:9]),rowMeans(annotated.tra[,10:12]),rowMeans(annotated.tra[,13:15]),rowMeans(annotated.tra[,16:18]),annotated.tra[,19:25])

colnames(annotated.tra.mean)[2:7]=names1

o = c(which(annotated.tra.mean[,12] %in% "SLC2A3"))

annotated.tra.mean.1=pivot_longer(annotated.tra.mean[o,], cols=2:7, names_to = "Stages" ,values_to="expression")

colnames(annotated.tra.mean.1)[1]= "transcript"

ggplot(annotated.tra.mean.1, aes(x = Stages, y =expression,group=as.character(transcript))) +
  geom_line(aes(color=transcript)) + scale_x_discrete(limits=names1) +
  labs(y=("Expression")) +
    ggtitle("Expression SLC2A3")
```

```{r Expression plot TEAD4}
names1 = c("1 cell stage", 
          "2 cell stage",
          "4 cell stage",
          "8 cell stage",
          "morula stage",
          "blastocyst stage")


annotated.tra.mean= cbind(rownames(annotated.tra),rowMeans(annotated.tra[,1:3]),rowMeans(annotated.tra[,4:6]),rowMeans(annotated.tra[,7:9]),rowMeans(annotated.tra[,10:12]),rowMeans(annotated.tra[,13:15]),rowMeans(annotated.tra[,16:18]),annotated.tra[,19:25])

colnames(annotated.tra.mean)[2:7]=names1

o = c(which(annotated.tra.mean[,12] %in% "TEAD4"))

annotated.tra.mean.1=pivot_longer(annotated.tra.mean[o,], cols=2:7, names_to = "Stages" ,values_to="expression")

colnames(annotated.tra.mean.1)[1]= "transcript"

ggplot(annotated.tra.mean.1, aes(x = Stages, y =expression,group=as.character(transcript))) +
  geom_line(aes(color=transcript)) + scale_x_discrete(limits=names1) +
  labs(y=("Expression")) +
    ggtitle("Expression TEAD4")
```

```{r Expression plot NQO1}
names1 = c("1 cell stage", 
          "2 cell stage",
          "4 cell stage",
          "8 cell stage",
          "morula stage",
          "blastocyst stage")


annotated.tra.mean= cbind(rownames(annotated.tra),rowMeans(annotated.tra[,1:3]),rowMeans(annotated.tra[,4:6]),rowMeans(annotated.tra[,7:9]),rowMeans(annotated.tra[,10:12]),rowMeans(annotated.tra[,13:15]),rowMeans(annotated.tra[,16:18]),annotated.tra[,19:25])

colnames(annotated.tra.mean)[2:7]=names1

o = c(which(annotated.tra.mean[,12] %in% "NQO1"))

annotated.tra.mean.1=pivot_longer(annotated.tra.mean[o,], cols=2:7, names_to = "Stages" ,values_to="expression")

colnames(annotated.tra.mean.1)[1]= "transcript"

ggplot(annotated.tra.mean.1, aes(x = Stages, y =expression,group=as.character(transcript))) +
  geom_line(aes(color=transcript)) + scale_x_discrete(limits=names1) +
  labs(y=("Expression")) +
    ggtitle("Expression NQO1")
```
### 3.4.3 Gene set enrichment analysis



```{r Annotate Expression set with Entrez Gene IDs}
# Convert Ensembl Gene IDs to Entrez IDs because Entrez IDs are needed for GSEA-code
# Extract unique transcripts in ensembl data
ensembl.unique = unique(ensembl.data$Transcript.stable.ID)
  
a = which(ensembl.unique %in% rownames(human.vsnrma.df2))
ensembl.new = ensembl.data[a,]
  
b = which(rownames(human.vsnrma.df2) %in% ensembl.unique)
human.vsnrma.df2.new = human.vsnrma.df2[b,]
  
ensembl.new = arrange(ensembl.new,Transcript.stable.ID)
human.vsnrma.df2.new = arrange(human.vsnrma.df2.new, rownames(human.vsnrma.df2.new))
  
annotated.human.vsnrma.df2 = cbind(human.vsnrma.df2.new, ensembl.new$Gene.stable.ID)
  
```

```{r Convert Ensembl Gene IDs to Entrez IDs}

#columns(org.Hs.eg.db) # returns list of available keytypes
annotated.human.vsnrma.df2$entrez = mapIds(org.Hs.eg.db,
                                             keys = annotated.human.vsnrma.df2$`ensembl.new$Gene.stable.ID`, #Column containing Ensembl gene ids
                                             column="ENTREZID",
                                             keytype="ENSEMBL")
```

```{r fit table with annotated expression set}
annotated.fit = lmFit(annotated.human.vsnrma.df2[,1:18], design)
```

```{r Create fit tables for every comparison}
fit.function = function(i){
    fit.table = contrasts.fit(annotated.fit, contrasts = cm[,i])
    fit.table = eBayes(fit.table)
  }
  
fit.1.2 = fit.function(1)
fit.1.4 = fit.function(2)
fit.1.8 = fit.function(3)
fit.1.m = fit.function(4)
fit.1.b = fit.function(5)
fit.2.4 = fit.function(6)
fit.2.8 = fit.function(7)
fit.2.m = fit.function(8)
fit.2.b = fit.function(9)
fit.4.8 = fit.function(10)
fit.4.m = fit.function(11)
fit.4.b = fit.function(12)
fit.8.m = fit.function(13)
fit.8.b = fit.function(14)
fit.m.b = fit.function(15)
```
  
  
```{r create a function to perform SEA on every fit table}
# GSEA with GO database
SEA.function.GO = function(x){
  enrich_go = goana(x, geneid = annotated.human.vsnrma.df2$entrez, species = "Hs")
  topGO(enrich_go, ontology ="BP",number=Inf)
}

setwd("~//documents//GitHub//2022-topic-04-team-03//Tables")
SEA.GO.1.8 = SEA.function.GO(fit.1.8)
write.csv(SEA.GO.1.8, file="SEA.GO.1.8.csv")
SEA.GO.8.m = SEA.function.GO(fit.8.m)
write.csv(SEA.GO.8.m, file="SEA.GO.8.m.csv")
SEA.GO.m.b = SEA.function.GO(fit.m.b)
write.csv(SEA.GO.m.b, file="SEA.GO.m.b.csv")
```

```{r}
organism = "org.Hs.eg.db"

# we want the log2 fold change 
GO.1.8.genes<- annotated.limma.1.8$logFC

# name the vector
names(GO.1.8.genes) <- annotated.limma.1.8$Gene.stable.ID


# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(GO.1.8.genes, decreasing = TRUE)

gse <- gseGO(geneList=gene_list, 
             ont ="BP", 
             keyType = "ENSEMBL", 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = organism, 
             pAdjustMethod = "none")
require(DOSE)
dotplot(gse, showCategory=10, split=".sign") + facet_grid(.~.sign)

```


### 3.4.4 Venn diagram

Using Venn diagram, we want to find out which genes are continuously up/down-regulated during every stage. They could be important for the embryogenesis.

```{r Venn diagram }
#creating demo data for Venn diagram
#1.8.tra Upregulated
tra.1.8.up=annotated.limma.1.8.tra[which(annotated.limma.1.8.tra$logFC>0),]$ensembl.transcript
#8.m.tra Upregulated
tra.8.m.up=annotated.limma.8.m.tra[which(annotated.limma.8.m.tra$logFC>0),]$ensembl.transcript
#m.b.tra Upregulated
tra.m.b.up=annotated.limma.m.b.tra[which(annotated.limma.m.b.tra$logFC>0),]$ensembl.transcript

list_venn.up = list(A = tra.1.8.up,B=tra.8.m.up,C=tra.m.b.up)

# Helper function to display Venn diagram.up
display_venn.up <- function(x){
  grid.newpage()
  venn_object <- venn.diagram(x, main=("Upregulated TRAS between different stages"), category.names=c("1-cell and 8-cell stage"," 8-cell and morula stage","morula and blastocyst stage"), fill = c("#009E73", "#E69F00", "#56B4E9"), filename= NULL)
  grid.draw(venn_object)
}

#venn plot up
display_venn.up(list_venn.up)

#create a function to give the name of the genes in the intersection of three stages
Intersect <- function (x) {  
  # Multiple set version of intersect
  # x is a list
  if (length(x) == 1) {
    unlist(x)
  } else if (length(x) == 2) {
    intersect(x[[1]], x[[2]])
  } else if (length(x) > 2){
    intersect(x[[1]], Intersect(x[-1]))
  }
}

intersect.tra.genes.up=Intersect(list_venn.up)
#9 genes are continuously up-regulated from 1-cell stage to blastocyst stage



annotated.tra[rownames(annotated.tra) %in% intersect.tra.genes.up,]


#1.8.tra downregulated
tra.1.8.down=annotated.limma.1.8.tra[which(annotated.limma.1.8.tra$logFC<0),]$ensembl.transcript
#8.m.tra downregulated
tra.8.m.down=annotated.limma.8.m.tra[which(annotated.limma.8.m.tra$logFC<0),]$ensembl.transcript
#m.b.tra downregulated
tra.m.b.down=annotated.limma.m.b.tra[which(annotated.limma.m.b.tra$logFC<0),]$ensembl.transcript

list_venn.down = list(A = tra.1.8.down,B=tra.8.m.down,C=tra.m.b.down)

# Helper function to display Venn diagram.down
display_venn.down <- function(x){
  grid.newpage()
  venn_object.down <- venn.diagram(x, main=("Downregulated TRAS between different stages"), cex.main=2, font.main=1, col.main= "red", category.names=c("1-cell and 8-cell stage"," 8-cell and morula stage","morula and blastocyst stage"),fill = c("#009E73", "#E69F00", "#56B4E9"), filename=NULL)
  grid.draw(venn_object.down)
}
display_venn.down(list_venn.down)

intersect.tra.genes.down=Intersect(list_venn.down)
#5 genes are continuously down-regulated from 1-cell stage to blastocyst stage

annotated.tra[rownames(annotated.tra) %in% intersect.tra.genes.down,]
```

### chemokine
```{r chemokine analysis}
#extract chemokine
chemokine.tra.list=annotated.tra[which(grepl("motif chemokine",annotated.tra[,24])),]
dim(chemokine.tra.list)
n_occur=data.frame(table(chemokine.tra.list$Chromosome.scaffold.name))
print(n_occur)
#62 25
chemokine.list=ensembl.data[which(grepl("motif chemokine",ensembl.data[,8])),]
dim(chemokine.list)
# 280 9

extract.chemokine.tra <- function(x){
x.chemokine.tra=x[which(rownames(x) %in% chemokine.tra.list$Transcript.stable.ID),]
  return(x.chemokine.tra)
}

annotated.limma.1.8.chemokine.tra=extract.chemokine.tra(annotated.limma.1.8)
dim(annotated.limma.1.8.chemokine.tra)
#0 9
annotated.limma.8.m.chemokine.tra=extract.chemokine.tra(annotated.limma.8.m)
dim(annotated.limma.8.m.chemokine.tra)
#0 9
annotated.limma.m.b.chemokine.tra=extract.chemokine.tra(annotated.limma.m.b)
dim(annotated.limma.m.b.chemokine.tra)
#0 9

extract.chemokine <- function(x){
  x.chemokine=x[which(rownames(x) %in% chemokine.list$Transcript.stable.ID),]
  return(x.chemokine)
}

annotated.limma.1.8.chemokine=extract.chemokine(annotated.limma.1.8)
dim(annotated.limma.1.8.chemokine)
#6 11

unique(annotated.limma.1.8.chemokine$gene.name)
#"CCR3" Down  "CXCL14" Up

annotated.limma.8.m.chemokine=extract.chemokine(annotated.limma.8.m)
dim(annotated.limma.8.m.chemokine)
#4 11

unique(annotated.limma.8.m.chemokine$gene.name)
#"CCL15" Up



annotated.limma.m.b.chemokine=extract.chemokine(annotated.limma.m.b)
dim(annotated.limma.m.b.chemokine)
#4 11



unique(annotated.limma.m.b.chemokine$gene.name)
#"CCL15" Down

human.vsnrma.df2$ensembl.transcripts=rownames(human.vsnrma.df2)
colnames(ensembl.data)[3]<- c("ensembl.transcripts")

human.ensembl=merge(x=human.vsnrma.df2,y=ensembl.data, all.x = TRUE,by="ensembl.transcripts")

human.ensembl.1.8.chemokine=human.ensembl[which(human.ensembl$ensembl.transcripts %in% rownames(annotated.limma.1.8.chemokine)),]

a=c("CCL15","CCR3","CXCL14")
human.ensembl.tra=human.ensembl[which(human.ensembl[,23] %in% a),]
human.ensembl.tra.mat=as.matrix(human.ensembl.tra[,2:19])
heatmap=pheatmap(human.ensembl.tra.mat, cluster_cols=FALSE, show_rownames = TRUE, legend=TRUE, fontsize_row=0.5)

```

